{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TpwdW_WDjemz"
      },
      "source": [
        "# Parsing and Pre-Processing GENIA Text Data"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AzpFTAjbjem1"
      },
      "source": [
        "The data is given in XML format, with the POS of each word identified. Biological entities are also labelled with the type of entity. We parse this data file and tag each word with the POS it is labelled as and also assign a label to it based on whether or not it is a biological entity."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2kw1cVz-msvX",
        "outputId": "34069612-713c-4ce6-a8ea-0f262a04bb17"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 15.4M  100 15.4M    0     0  31.7M      0 --:--:-- --:--:-- --:--:-- 31.7M\n"
          ]
        }
      ],
      "source": [
        "!curl -o 'data' https://raw.githubusercontent.com/Shkev/Biomedical-Named-Entity-Recognition-SVM/main/data/GENIAcorpus3.02.merged.xml"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "r7prSj2Yjem2"
      },
      "outputs": [],
      "source": [
        "import bs4\n",
        "from bs4 import BeautifulSoup\n",
        "from tqdm.notebook import tqdm\n",
        "import numpy as np\n",
        "from typing import List, Dict, Tuple, Set\n",
        "\n",
        "data_file_path = '/content/data'"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fG_5XB0Qjem3"
      },
      "source": [
        "## Extracting Words and Named Entities from Texts\n",
        "\n",
        "Here we parse the XML data file to get the POS of each word and determine the semantic biological meaning of each word (if it has one). The words with a biological meaning will be labelled as a 1, and the non-biological words will be labelled with a -1."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "pm65onwDjem3"
      },
      "outputs": [],
      "source": [
        "connectors = ('(AND ', '(OR ', '(AND/OR ', '(TO ', '(BUT_NOT ', '(AS_WELL_AS ', '(VERSUS ', '(NEITHER_NOR ', '(THAN ', '(AND_NOT ', '(NOT_ONLY_BUT_ALSO ')\n",
        "def parse_name_connectors(text: str) -> List[str]:\n",
        "    \"\"\"\n",
        "    Parses through nested connectors in labelled entities and returns the separated entities\n",
        "\n",
        "    Args:\n",
        "        text (str): The string to be parsed for individual entities\n",
        "      \n",
        "    Returns:\n",
        "        A list containing all entities that are contained in the given text string\n",
        "    \"\"\"\n",
        "    if text.startswith(connectors):\n",
        "        # we can just split on spaces bc spaces are replaced by '_' in all named entity entries\n",
        "        # so splitting won't split the names of the entities\n",
        "        components = text.split()\n",
        "        components = [comp for comp in components if comp+' ' not in connectors]\n",
        "        components = [comp[:-1] if comp.endswith(')') else comp for comp in components]\n",
        "        return components\n",
        "    return [text.strip()]\n",
        "\n",
        "def get_words_part_of_speech(sentence: bs4.element.Tag, words_pos_dict: Dict[str, Set[str]]) -> None:\n",
        "    \"\"\"\n",
        "    Extracts words in given sentence and their respective part of speech.\n",
        "    Updates words_pos_dict (passed by reference) to contain these words and their POS.\n",
        "\n",
        "    Arg:\n",
        "        sentence (bs4.element.Tag): The sentence to extract the words and POS from.\n",
        "        words_pos_dict (dict): The dictionary to update with the words and POS.\n",
        "    \"\"\"\n",
        "    words = sentence.find_all('w')\n",
        "    fragments = []\n",
        "    for word in words:\n",
        "        text = word.text.strip()\n",
        "\n",
        "        if word['c'] == '*':\n",
        "            fragments.append(text)\n",
        "            continue\n",
        "        fragments.append(text)\n",
        "        text = ''.join(fragments)\n",
        "        fragments.clear()\n",
        "\n",
        "        if text not in words_pos_dict:\n",
        "            words_pos_dict[text] = set()\n",
        "        # words that can be multiple POS indicated with '|' between each POS tag\n",
        "        pos = word['c'].split('|')\n",
        "        words_pos_dict[text].update(pos)\n",
        "\n",
        "def get_named_bio_entities(sentence: bs4.element.Tag, named_bio_entities: Set[str]) -> None:\n",
        "    \"\"\"\n",
        "    Extracts named biological entities in given sentence.\n",
        "    Updates named_bio_entities (passed by reference) to contain these entities.\n",
        "\n",
        "    Arg:\n",
        "        sentence (bs4.element.Tag): The sentence to extract the named biological entities from.\n",
        "        named_bio_entities (set): The dictionary to update with the named biological entities.\n",
        "    \"\"\"\n",
        "    bio_entities = sentence.find_all('cons')\n",
        "    for ent in bio_entities:\n",
        "        if ent.has_attr('lex'):\n",
        "            text = ent['lex'].strip()\n",
        "            # if it has semantic meaning labelled it is a bio named entity\n",
        "            if ent.has_attr('sem'):\n",
        "                text = parse_name_connectors(text)\n",
        "                named_bio_entities.update(text)\n",
        "        else:\n",
        "            print(\"This constituent has no text: \", ent)\n",
        "\n",
        "def extract_data_from_xml(data_file_path: str) -> Tuple[Dict[str, Set[str]], Set[str]]:\n",
        "    \"\"\"\n",
        "    Extracts the data from the xml file and returns a list of tuples\n",
        "    containing the text and the label.\n",
        "    Note that punctuation marks are kept in the data as was done in the presented paper.\n",
        "    However, they are not attached to any of the words.\n",
        "    \n",
        "    Args:\n",
        "        data_file_path (str): The path to the xml file.\n",
        "        \n",
        "    Returns:\n",
        "        A dictionary containing all words and their POS and \n",
        "        a set of all biomedical named entities in the texts.\n",
        "    \"\"\"\n",
        "    data = []\n",
        "    words_pos_dict = dict() # dict of set\n",
        "    named_bio_entities = set()\n",
        "    with open(data_file_path, \"r\") as f:\n",
        "        xml = f.read()\n",
        "        print(\"Parsing XML file with BeautifulSoup...\")\n",
        "        soup = BeautifulSoup(xml, 'xml')\n",
        "        print(\"Parsing Done. Now extracting info...\")\n",
        "        articles = soup.find_all('article')\n",
        "        for art in tqdm(articles):\n",
        "            sentences = art.find_all('sentence')\n",
        "            for sent in sentences:\n",
        "                get_words_part_of_speech(sent, words_pos_dict)\n",
        "                get_named_bio_entities(sent, named_bio_entities)\n",
        "    print(\"Done\")\n",
        "    return words_pos_dict, named_bio_entities\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "referenced_widgets": [
            "416e4d9a63ab48818f451146351e3241",
            "1c82c6e6139a4717a16323eafde603de",
            "6ee5b482980c48948be2666771a47d83",
            "e5659618745e42469a1722dbc639f058",
            "cb317891831f453a97168a4bc71b51a5",
            "da8006e95e1d4c3ab4cb6062ebb2dd1e",
            "ae5d5fcaa6944a528c08504872a666b2",
            "87cc8d887f504979b98bf467585f4364",
            "2684f161f67d463cb1dd9a539d14d6d9",
            "90d8ae28b880486cbff0f062dfc35174",
            "9e93286e84b34eac9900e7ceca8923be"
          ],
          "height": 0
        },
        "id": "8jurbRg_jem4",
        "outputId": "d01538aa-9f6a-4122-d3fc-bcc8f3982db0"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Parsing XML file with BeautifulSoup...\n",
            "Parsing Done. Now extracting info...\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "  0%|          | 0/2000 [00:00<?, ?it/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "416e4d9a63ab48818f451146351e3241"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "This constituent has no text:  <cons sem=\"G#cell_line\"><cons sem=\"G#cell_line\"><w c=\"NN\">THP-1</w></cons> <w c=\"JJ\">mononuclear</w> <w c=\"NN\">phagocyte</w> <w c=\"NN\">cell</w> <w c=\"NN\">line</w></cons>\n",
            "This constituent has no text:  <cons sem=\"G#cell_line\"><w c=\"NN\">THP-1</w></cons>\n",
            "Done\n"
          ]
        }
      ],
      "source": [
        "words_pos_dict, named_bio_ent = extract_data_from_xml(data_file_path)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GvdU2ecgdl1f"
      },
      "source": [
        "The paper explains that they label each named entity with their POS that they extract from the dataset. However, the dataset only assigns POS tags to single words and not multi-word entities. I assume that the writers of the paper limited their view to only the single word biological named entities to be able to accurately and efficiently determine their POS. There may be a way to determine the POS of several words (maybe using external libraries) but I am not aware of any of these methods, so this is how I will proceed. Also, many of the multiword entities have single word entities nested in them, which can in some cases represent the entire entity for the purposes of the data. "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dWH7mOW1jem5",
        "outputId": "5becd2dc-091f-4ac4-c5b6-616ce2ac5150"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "4805"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ],
      "source": [
        "single_word_bio_ent = named_bio_ent.intersection(set(words_pos_dict.keys()))\n",
        "len(single_word_bio_ent)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "faJWHGIjlcT3"
      },
      "source": [
        "## Labels\n",
        "\n",
        "Each data point is assigned a 1 if it is a biological entity and a -1 if it is not."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "Y_jE6iz3zZfh"
      },
      "outputs": [],
      "source": [
        "# maps each word to its label\n",
        "word_labels = {word : 1 if word in single_word_bio_ent else -1 for word in words_pos_dict.keys()}"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "t6EYBiwNfKIZ"
      },
      "source": [
        "## Features\n",
        "\n",
        "We use the extracted data (POS of named entities and the entities themselves) to construct the feature vectors for each word which will be given to the model to train on.\n",
        "\n",
        "As described in the paper, we use both the POS of the words as well as structural features for the words."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "VScKEgDo1yrr"
      },
      "outputs": [],
      "source": [
        "# stored as dict of dict. Maps word to feature dict\n",
        "# for each sub-dict, key is index of feature and value is 1\n",
        "# features that word does not have are excluded (all values in the dicts should be 1)\n",
        "word_features = dict()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UQ4wLfZKlSB_"
      },
      "source": [
        "### POS Features"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "yiQisoOhh1gv"
      },
      "outputs": [],
      "source": [
        "pos_tags = set().union(*words_pos_dict.values())\n",
        "num_pos_feat = len(pos_tags)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "55welwVqkf8e"
      },
      "source": [
        "Not sure why, but there are more POS tags in this dataset than the study said they extracted from theirs. This may be due to a discrepency in the version of the dataset being used (version is not specified in the paper).\n",
        "\n",
        "A unique index is assigned to each POS tag. The value of the index does not matter as long as it is consistent throughout the rest of the code.\n",
        "\n",
        "These indices are used to assign features to each word. If a word is a certain POS, the entry `idx:1` is added to its feature dict (where idx is the index corresponding to the POS tag)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4lmU_BO4fg9z",
        "outputId": "886c63dd-b593-4f7c-cb9d-0c16725a05a2"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'JJ': 0,\n",
              " 'NN': 1,\n",
              " '': 2,\n",
              " 'NNP': 3,\n",
              " 'VBG': 4,\n",
              " 'TO': 5,\n",
              " 'POS': 6,\n",
              " 'N': 7,\n",
              " 'RBS': 8,\n",
              " \"''\": 9,\n",
              " '-': 10,\n",
              " 'WDT': 11,\n",
              " 'VBZ': 12,\n",
              " 'CT': 13,\n",
              " 'WP$': 14,\n",
              " 'PDT': 15,\n",
              " 'SYM': 16,\n",
              " 'WP': 17,\n",
              " 'FW': 18,\n",
              " 'VBD': 19,\n",
              " 'XT': 20,\n",
              " ',': 21,\n",
              " 'PRP$': 22,\n",
              " 'IN': 23,\n",
              " 'DT': 24,\n",
              " 'RBR': 25,\n",
              " 'WRB': 26,\n",
              " 'JJR': 27,\n",
              " 'NNPS': 28,\n",
              " 'PRP': 29,\n",
              " 'RB': 30,\n",
              " '``': 31,\n",
              " '(': 32,\n",
              " 'NNS': 33,\n",
              " 'EX': 34,\n",
              " 'VBN': 35,\n",
              " 'CC': 36,\n",
              " '.': 37,\n",
              " 'VBP': 38,\n",
              " 'PP': 39,\n",
              " 'RP': 40,\n",
              " 'JJS': 41,\n",
              " 'CD': 42,\n",
              " 'MD': 43,\n",
              " 'VB': 44,\n",
              " ':': 45,\n",
              " 'LS': 46,\n",
              " ')': 47}"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ],
      "source": [
        "pos_indices = {pos : idx for (idx, pos) in enumerate(pos_tags)}\n",
        "pos_indices"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rIo3dAYXlSNA",
        "outputId": "81c76842-ea4a-4440-c759-84466d2ad8e6"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'IL-2': {1: 1},\n",
              " 'gene': {1: 1},\n",
              " 'expression': {1: 1},\n",
              " 'and': {36: 1},\n",
              " 'NF-kappa': {1: 1},\n",
              " 'B': {1: 1},\n",
              " 'activation': {1: 1},\n",
              " 'through': {23: 1},\n",
              " 'CD28': {1: 1},\n",
              " 'requires': {12: 1},\n",
              " 'reactive': {0: 1},\n",
              " 'oxygen': {1: 1},\n",
              " 'production': {1: 1},\n",
              " 'by': {23: 1},\n",
              " '5-lipoxygenase': {0: 1, 1: 1},\n",
              " '.': {37: 1},\n",
              " 'Activation': {1: 1},\n",
              " 'of': {23: 1},\n",
              " 'the': {24: 1, 13: 1},\n",
              " 'surface': {1: 1},\n",
              " 'receptor': {1: 1},\n",
              " 'provides': {12: 1},\n",
              " 'a': {1: 1, 20: 1, 4: 1, 24: 1, 46: 1, 13: 1},\n",
              " 'major': {0: 1},\n",
              " 'costimulatory': {0: 1, 1: 1},\n",
              " 'signal': {0: 1, 1: 1, 44: 1, 38: 1},\n",
              " 'for': {23: 1},\n",
              " 'T': {0: 1, 1: 1},\n",
              " 'cell': {1: 1},\n",
              " 'resulting': {4: 1},\n",
              " 'in': {18: 1, 1: 1, 23: 1, 30: 1, 40: 1},\n",
              " 'enhanced': {19: 1, 0: 1, 35: 1},\n",
              " 'interleukin-2': {1: 1},\n",
              " '(': {32: 1},\n",
              " ')': {2: 1, 47: 1},\n",
              " 'proliferation': {1: 1},\n",
              " 'In': {18: 1, 23: 1},\n",
              " 'primary': {0: 1},\n",
              " 'lymphocytes': {33: 1},\n",
              " 'we': {29: 1},\n",
              " 'show': {44: 1, 38: 1},\n",
              " 'that': {23: 1, 24: 1, 11: 1},\n",
              " 'ligation': {1: 1},\n",
              " 'leads': {12: 1},\n",
              " 'to': {5: 1},\n",
              " 'rapid': {0: 1},\n",
              " 'intracellular': {0: 1},\n",
              " 'formation': {1: 1},\n",
              " 'intermediates': {33: 1},\n",
              " 'ROIs': {33: 1},\n",
              " 'which': {11: 1, 17: 1},\n",
              " 'are': {38: 1},\n",
              " 'required': {19: 1, 0: 1, 35: 1},\n",
              " 'CD28-mediated': {0: 1},\n",
              " 'B/CD28-responsive': {0: 1},\n",
              " 'complex': {0: 1, 1: 1},\n",
              " 'Delineation': {1: 1},\n",
              " 'signaling': {0: 1, 1: 1, 4: 1},\n",
              " 'cascade': {1: 1},\n",
              " 'was': {19: 1},\n",
              " 'found': {19: 1, 35: 1},\n",
              " 'involve': {44: 1, 38: 1},\n",
              " 'protein': {1: 1},\n",
              " 'tyrosine': {1: 1},\n",
              " 'kinase': {1: 1},\n",
              " 'activity': {1: 1},\n",
              " ',': {37: 1, 21: 1},\n",
              " 'followed': {19: 1, 35: 1},\n",
              " 'phospholipase': {1: 1},\n",
              " 'A2': {1: 1},\n",
              " 'Our': {22: 1},\n",
              " 'data': {33: 1, 1: 1},\n",
              " 'suggest': {44: 1, 38: 1},\n",
              " 'lipoxygenase': {1: 1},\n",
              " 'metabolites': {33: 1},\n",
              " 'activate': {44: 1, 38: 1},\n",
              " 'ROI': {33: 1, 1: 1},\n",
              " 'then': {30: 1},\n",
              " 'induce': {44: 1, 38: 1},\n",
              " 'via': {23: 1},\n",
              " 'These': {24: 1},\n",
              " 'findings': {33: 1},\n",
              " 'should': {43: 1},\n",
              " 'be': {44: 1},\n",
              " 'useful': {0: 1},\n",
              " 'therapeutic': {0: 1},\n",
              " 'strategies': {33: 1},\n",
              " 'development': {1: 1},\n",
              " 'immunosuppressants': {33: 1},\n",
              " 'targeting': {1: 1, 4: 1},\n",
              " 'pathway': {1: 1},\n",
              " 'The': {24: 1, 3: 1},\n",
              " 'peri-kappa': {1: 1},\n",
              " 'site': {1: 1},\n",
              " 'mediates': {12: 1},\n",
              " 'human': {0: 1, 1: 1},\n",
              " 'immunodeficiency': {1: 1},\n",
              " 'virus': {1: 1},\n",
              " 'type': {1: 1},\n",
              " '2': {42: 1, 46: 1},\n",
              " 'enhancer': {1: 1},\n",
              " 'monocytes': {33: 1},\n",
              " 'but': {36: 1},\n",
              " 'not': {30: 1},\n",
              " 'cells': {33: 1},\n",
              " 'Human': {0: 1},\n",
              " 'HIV-2': {1: 1},\n",
              " 'like': {0: 1, 23: 1},\n",
              " 'HIV-1': {1: 1},\n",
              " 'causes': {33: 1, 12: 1},\n",
              " 'AIDS': {1: 1, 3: 1},\n",
              " 'is': {19: 1, 12: 1},\n",
              " 'associated': {19: 1, 0: 1, 35: 1},\n",
              " 'with': {23: 1},\n",
              " 'cases': {33: 1},\n",
              " 'primarily': {30: 1},\n",
              " 'West': {0: 1, 3: 1},\n",
              " 'Africa': {3: 1},\n",
              " 'display': {1: 1, 44: 1, 38: 1},\n",
              " 'significant': {0: 1},\n",
              " 'differences': {33: 1},\n",
              " 'nucleic': {0: 1},\n",
              " 'acid': {0: 1, 1: 1},\n",
              " 'sequence': {1: 1},\n",
              " 'natural': {0: 1},\n",
              " 'history': {1: 1},\n",
              " 'clinical': {0: 1},\n",
              " 'disease': {1: 1},\n",
              " 'Consistent': {0: 1},\n",
              " 'these': {24: 1},\n",
              " 'have': {44: 1, 38: 1},\n",
              " 'previously': {30: 1},\n",
              " 'demonstrated': {19: 1, 35: 1},\n",
              " 'enhancer/promoter': {0: 1, 1: 1},\n",
              " 'region': {1: 1},\n",
              " 'functions': {33: 1, 12: 1},\n",
              " 'quite': {30: 1},\n",
              " 'differently': {30: 1},\n",
              " 'from': {23: 1},\n",
              " 'Whereas': {23: 1},\n",
              " 'following': {0: 1, 1: 1, 4: 1},\n",
              " 'T-cell': {1: 1},\n",
              " 'stimulation': {1: 1},\n",
              " 'mediated': {19: 1, 0: 1, 35: 1},\n",
              " 'largely': {30: 1},\n",
              " 'binding': {0: 1, 1: 1, 4: 1},\n",
              " 'transcription': {1: 1},\n",
              " 'factor': {1: 1, 23: 1},\n",
              " 'two': {42: 1},\n",
              " 'adjacent': {0: 1},\n",
              " 'kappa': {1: 1},\n",
              " 'sites': {33: 1},\n",
              " 'long': {0: 1, 30: 1},\n",
              " 'terminal': {0: 1, 1: 1},\n",
              " 'repeat': {1: 1},\n",
              " 'dependent': {0: 1},\n",
              " 'on': {23: 1},\n",
              " 'four': {42: 1},\n",
              " 'cis-acting': {0: 1, 1: 1, 4: 1},\n",
              " 'elements': {33: 1},\n",
              " ':': {45: 1},\n",
              " 'single': {0: 1},\n",
              " 'purine-rich': {0: 1},\n",
              " 'PuB1': {1: 1},\n",
              " 'PuB2': {1: 1},\n",
              " 'pets': {1: 1},\n",
              " 'We': {29: 1},\n",
              " 'now': {1: 1, 30: 1},\n",
              " 'identified': {19: 1, 35: 1},\n",
              " 'novel': {0: 1},\n",
              " 'element': {1: 1},\n",
              " 'within': {23: 1},\n",
              " 'immediately': {30: 1},\n",
              " 'upstream': {0: 1, 30: 1},\n",
              " 'designated': {19: 1, 0: 1, 35: 1},\n",
              " 'This': {24: 1, 13: 1},\n",
              " 'conserved': {0: 1, 19: 1, 35: 1},\n",
              " 'among': {23: 1},\n",
              " 'isolates': {33: 1},\n",
              " 'closely': {30: 1},\n",
              " 'related': {0: 1, 35: 1},\n",
              " 'simian': {0: 1},\n",
              " 'transfection': {1: 1},\n",
              " 'assays': {33: 1, 1: 1},\n",
              " 'this': {24: 1},\n",
              " 'mediate': {44: 1, 38: 1},\n",
              " 'monocytic': {0: 1},\n",
              " 'lines': {33: 1},\n",
              " 'first': {0: 1, 30: 1},\n",
              " 'description': {1: 1},\n",
              " 'an': {24: 1, 13: 1},\n",
              " 'displays': {12: 1},\n",
              " 'such': {0: 1, 23: 1, 15: 1},\n",
              " 'monocyte': {0: 1, 1: 1},\n",
              " 'specificity': {1: 1},\n",
              " 'no': {24: 1, 30: 1},\n",
              " 'comparable': {0: 1},\n",
              " 'has': {12: 1},\n",
              " 'been': {35: 1},\n",
              " 'clearly': {30: 1},\n",
              " 'defined': {19: 1, 35: 1},\n",
              " 'While': {23: 1},\n",
              " 'nuclear': {0: 1},\n",
              " 's': {33: 1, 1: 1, 12: 1},\n",
              " 'both': {24: 1, 30: 1, 36: 1, 15: 1},\n",
              " 'peripheral': {0: 1},\n",
              " 'blood': {1: 1},\n",
              " 'binds': {12: 1},\n",
              " 'electrophoretic': {0: 1},\n",
              " 'mobility': {1: 1},\n",
              " 'shift': {1: 1, 44: 1},\n",
              " 'either': {24: 1, 30: 1, 36: 1},\n",
              " 'different': {0: 1},\n",
              " 'versus': {36: 1},\n",
              " 'or': {36: 1},\n",
              " 'recognizing': {4: 1},\n",
              " 'undergoes': {12: 1},\n",
              " 'differential': {0: 1},\n",
              " 'modification': {1: 1},\n",
              " 'thus': {30: 1},\n",
              " 'supporting': {1: 1, 4: 1},\n",
              " 'Further': {0: 1, 23: 1, 30: 1},\n",
              " 'while': {23: 1},\n",
              " 'specific': {0: 1},\n",
              " 'constitutive': {0: 1},\n",
              " 'seen': {35: 1},\n",
              " 'phorbol': {1: 1},\n",
              " 'esters': {33: 1},\n",
              " 'induces': {12: 1},\n",
              " 'additional': {0: 1},\n",
              " 'Understanding': {4: 1},\n",
              " 'monocyte-specific': {0: 1},\n",
              " 'function': {1: 1, 44: 1, 38: 1},\n",
              " 'may': {1: 1, 43: 1},\n",
              " 'ultimately': {30: 1},\n",
              " 'provide': {44: 1, 38: 1},\n",
              " 'insight': {1: 1},\n",
              " 'into': {23: 1},\n",
              " 'role': {1: 1},\n",
              " 'play': {1: 1, 44: 1, 38: 1},\n",
              " 'HIV': {1: 1, 38: 1},\n",
              " 'pathogenesis': {1: 1},\n",
              " 'E1A': {1: 1},\n",
              " 'susceptibility': {1: 1},\n",
              " 'killing': {1: 1, 4: 1},\n",
              " 'NK': {1: 1},\n",
              " 'immortalization': {1: 1},\n",
              " 'adenovirus': {1: 1},\n",
              " 'infection': {1: 1},\n",
              " 'Adenovirus': {1: 1},\n",
              " 'Ad': {1: 1},\n",
              " 'were': {19: 1},\n",
              " 'used': {19: 1, 35: 1},\n",
              " 'model': {0: 1, 1: 1, 44: 1},\n",
              " 'changes': {33: 1, 12: 1},\n",
              " 'caused': {19: 1, 35: 1},\n",
              " 'transient': {0: 1, 1: 1},\n",
              " 'vs': {36: 1},\n",
              " 'stable': {0: 1},\n",
              " 'Only': {30: 1},\n",
              " 'stably': {30: 1},\n",
              " 'transfected': {19: 1, 0: 1, 35: 1},\n",
              " 'target': {1: 1, 44: 1, 38: 1},\n",
              " 'exhibited': {19: 1, 35: 1},\n",
              " 'cytolytic': {0: 1},\n",
              " 'despite': {23: 1},\n",
              " 'equivalent': {0: 1, 1: 1},\n",
              " 'levels': {33: 1},\n",
              " 'proteins': {33: 1},\n",
              " 'Ad-infected': {0: 1},\n",
              " 'targets': {33: 1, 12: 1},\n",
              " 'inability': {1: 1},\n",
              " 'products': {33: 1},\n",
              " 'during': {23: 1},\n",
              " 'explained': {35: 1},\n",
              " 'inhibitory': {0: 1},\n",
              " 'effect': {1: 1, 44: 1},\n",
              " 'viral': {0: 1},\n",
              " 'otherwise': {0: 1, 30: 1},\n",
              " 'susceptible': {0: 1},\n",
              " 'effects': {33: 1},\n",
              " 'class': {1: 1},\n",
              " 'I': {1: 1, 46: 1, 42: 1, 29: 1},\n",
              " 'MHC': {1: 1},\n",
              " 'antigen': {0: 1, 1: 1},\n",
              " 'phenotypes': {33: 1},\n",
              " 'infected': {0: 1, 35: 1},\n",
              " 'suggests': {12: 1},\n",
              " 'effective': {0: 1},\n",
              " 'immunologic': {0: 1},\n",
              " 'barrier': {1: 1},\n",
              " 'against': {23: 1},\n",
              " 'vivo': {18: 1},\n",
              " 'survival': {0: 1, 1: 1},\n",
              " 'neoplastic': {0: 1},\n",
              " 'progression': {1: 1},\n",
              " 'E1A-immortalized': {0: 1},\n",
              " 'emerge': {44: 1, 38: 1},\n",
              " 'reservoir': {1: 1},\n",
              " 'persistently': {30: 1},\n",
              " 'host': {1: 1},\n",
              " 'Distinct': {0: 1},\n",
              " 'properties': {33: 1},\n",
              " 'identify': {44: 1, 38: 1},\n",
              " 'functionally': {30: 1},\n",
              " 'CD4': {1: 1},\n",
              " 'epitopes': {33: 1},\n",
              " 'coreceptor': {1: 1},\n",
              " 'interacts': {12: 1},\n",
              " 'non-polymorphic': {0: 1},\n",
              " 'regions': {33: 1},\n",
              " 'histocompatibility': {1: 1},\n",
              " 'II': {42: 1},\n",
              " 'molecules': {33: 1},\n",
              " 'antigen-presenting': {0: 1},\n",
              " 'contributes': {12: 1},\n",
              " 'investigated': {19: 1, 35: 1},\n",
              " 'triggering': {1: 1, 4: 1},\n",
              " 'activating': {0: 1, 1: 1, 4: 1},\n",
              " 'signals': {33: 1, 12: 1},\n",
              " 'lymphoma': {1: 1},\n",
              " 'using': {4: 1},\n",
              " 'monoclonal': {0: 1},\n",
              " 'antibodies': {33: 1},\n",
              " 'mAb': {1: 1},\n",
              " 'recognize': {44: 1, 38: 1},\n",
              " 'demonstrate': {44: 1, 38: 1},\n",
              " 'delivers': {12: 1},\n",
              " 'capable': {0: 1},\n",
              " 'NF-AT': {1: 1},\n",
              " 'anti-CD4': {0: 1},\n",
              " 'gp120': {1: 1},\n",
              " 'could': {43: 1},\n",
              " 'all': {24: 1, 15: 1},\n",
              " 'trigger': {1: 1, 44: 1, 38: 1},\n",
              " 'kinases': {33: 1},\n",
              " 'p56lck': {1: 1},\n",
              " 'p59fyn': {1: 1},\n",
              " 'phosphorylation': {1: 1},\n",
              " 'Shc': {1: 1},\n",
              " 'adaptor': {1: 1},\n",
              " 'Ras': {33: 1, 1: 1},\n",
              " 'they': {29: 1},\n",
              " 'differed': {19: 1, 35: 1},\n",
              " 'significantly': {30: 1},\n",
              " 'their': {22: 1},\n",
              " 'ability': {1: 1},\n",
              " 'Lack': {1: 1},\n",
              " 'full': {0: 1},\n",
              " 'correlated': {19: 1, 0: 1, 35: 1},\n",
              " 'dramatically': {30: 1},\n",
              " 'reduced': {19: 1, 0: 1, 35: 1},\n",
              " 'capacity': {1: 1},\n",
              " 'calcium': {1: 1},\n",
              " 'flux': {1: 1},\n",
              " 'complemented': {19: 1, 35: 1},\n",
              " 'ionophore': {1: 1},\n",
              " 'results': {33: 1, 12: 1},\n",
              " 'distinct': {0: 1},\n",
              " 'involved': {19: 1, 0: 1, 35: 1},\n",
              " 'Ras/protein': {1: 1},\n",
              " 'C': {1: 1},\n",
              " 'pathways': {33: 1},\n",
              " 'Ligand-dependent': {0: 1},\n",
              " 'repression': {1: 1},\n",
              " 'erythroid': {0: 1},\n",
              " 'GATA-1': {1: 1},\n",
              " 'estrogen': {1: 1},\n",
              " 'High-dose': {0: 1},\n",
              " 'administration': {1: 1},\n",
              " 'anemia': {1: 1},\n",
              " 'mammals': {33: 1},\n",
              " 'chickens': {33: 1},\n",
              " 'estrogens': {33: 1},\n",
              " 'stimulate': {44: 1, 38: 1},\n",
              " 'outgrowth': {1: 1},\n",
              " 'bone': {1: 1},\n",
              " 'marrow-derived': {0: 1},\n",
              " 'progenitor': {1: 1},\n",
              " 'delay': {1: 1, 44: 1, 38: 1},\n",
              " 'maturation': {1: 1},\n",
              " 'down-regulation': {1: 1},\n",
              " 'many': {0: 1},\n",
              " 'cell-specific': {0: 1},\n",
              " 'genes': {33: 1},\n",
              " 'including': {4: 1},\n",
              " 'alpha-': {1: 1},\n",
              " 'beta-globin': {0: 1, 1: 1, 23: 1},\n",
              " 'band': {1: 1},\n",
              " '3': {42: 1, 46: 1},\n",
              " '4.1': {42: 1},\n",
              " 'histone': {1: 1},\n",
              " 'H5': {1: 1},\n",
              " 'here': {30: 1},\n",
              " 'also': {30: 1},\n",
              " 'reduce': {44: 1, 38: 1},\n",
              " 'number': {1: 1},\n",
              " 'marrow': {0: 1, 1: 1},\n",
              " 'cultures': {33: 1},\n",
              " 'To': {5: 1},\n",
              " 'address': {44: 1, 38: 1},\n",
              " 'potential': {0: 1, 1: 1},\n",
              " 'mechanisms': {33: 1},\n",
              " 'suppress': {44: 1, 38: 1},\n",
              " 'erythropoiesis': {1: 1},\n",
              " 'examined': {19: 1, 35: 1},\n",
              " 'participates': {12: 1},\n",
              " 'regulation': {1: 1},\n",
              " 'majority': {1: 1},\n",
              " 'necessary': {0: 1},\n",
              " 'erythrocytes': {33: 1},\n",
              " 'transcriptional': {0: 1},\n",
              " 'strongly': {30: 1},\n",
              " 'repressed': {19: 1, 35: 1},\n",
              " 'ER': {1: 1},\n",
              " 'ligand-dependent': {0: 1},\n",
              " 'manner': {1: 1},\n",
              " 'reversible': {0: 1},\n",
              " 'presence': {1: 1},\n",
              " '4-hydroxytamoxifen': {1: 1},\n",
              " 'ER-mediated': {0: 1},\n",
              " 'occurs': {12: 1},\n",
              " 'artificial': {0: 1},\n",
              " 'promoter': {0: 1, 1: 1},\n",
              " 'containing': {1: 1, 4: 1},\n",
              " 'GATA-binding': {0: 1, 1: 1},\n",
              " 'as': {23: 1, 30: 1},\n",
              " 'well': {1: 1, 30: 1},\n",
              " 'context': {1: 1},\n",
              " 'intact': {0: 1},\n",
              " 'normally': {30: 1},\n",
              " 'regulated': {19: 1, 0: 1, 35: 1},\n",
              " 'bind': {44: 1, 38: 1},\n",
              " 'each': {24: 1},\n",
              " 'other': {0: 1, 24: 1},\n",
              " 'vitro': {18: 1},\n",
              " 'absence': {1: 1},\n",
              " 'DNA': {1: 1},\n",
              " 'coimmunoprecipitation': {1: 1},\n",
              " 'experiments': {33: 1},\n",
              " 'COS': {1: 1},\n",
              " 'associate': {44: 1, 38: 1},\n",
              " 'Mapping': {1: 1},\n",
              " 'indicate': {44: 1, 38: 1},\n",
              " 'form': {1: 1, 23: 1, 44: 1, 38: 1},\n",
              " 'at': {23: 1},\n",
              " 'least': {41: 1, 8: 1},\n",
              " 'contacts': {33: 1},\n",
              " 'finger': {1: 1},\n",
              " 'N-terminal': {0: 1},\n",
              " 'domain': {0: 1, 1: 1},\n",
              " 'speculate': {44: 1, 38: 1},\n",
              " 'exert': {44: 1, 38: 1, 35: 1},\n",
              " 'modulating': {0: 1, 1: 1, 4: 1},\n",
              " 'protein-protein': {0: 1},\n",
              " 'interaction': {1: 1},\n",
              " 'ABSTRACT': {1: 1},\n",
              " 'TRUNCATED': {35: 1},\n",
              " 'AT': {1: 1, 23: 1},\n",
              " '250': {42: 1},\n",
              " 'WORDS': {33: 1},\n",
              " 'Mouse': {1: 1},\n",
              " 'alpha': {1: 1},\n",
              " 'Interleukin-1': {1: 1},\n",
              " 'control': {0: 1, 1: 1, 44: 1, 38: 1},\n",
              " 'shown': {35: 1},\n",
              " 'interleukin-1': {1: 1},\n",
              " 'IL-1': {1: 1},\n",
              " 'IL-2R': {1: 1},\n",
              " 'CD4-CD8- murine': {0: 1},\n",
              " 'lymphocyte': {0: 1, 1: 1},\n",
              " 'precursors': {33: 1},\n",
              " 'Here': {30: 1},\n",
              " 'map': {1: 1, 44: 1, 38: 1},\n",
              " 'interleukin': {1: 1},\n",
              " 'responsiveness': {1: 1},\n",
              " 'mouse': {1: 1},\n",
              " 'thymic': {0: 1},\n",
              " 'lymphoma-derived': {0: 1},\n",
              " 'hybridoma': {1: 1},\n",
              " 'PC60': {1: 1},\n",
              " 'response': {1: 1, 38: 1},\n",
              " '+': {0: 1, 36: 1},\n",
              " 'biphasic': {0: 1},\n",
              " 'synthesis-independent': {0: 1},\n",
              " 'appearance': {1: 1},\n",
              " 'mRNA': {1: 1},\n",
              " 'blocked': {19: 1, 35: 1},\n",
              " 'inhibitors': {33: 1},\n",
              " 'It': {29: 1},\n",
              " 'primes': {12: 1},\n",
              " 'become': {44: 1, 38: 1, 35: 1},\n",
              " 'responsive': {0: 1, 1: 1},\n",
              " 'thereby': {30: 1},\n",
              " 'prepares': {12: 1},\n",
              " 'second': {0: 1, 1: 1, 30: 1},\n",
              " 'phase': {1: 1},\n",
              " '100-fold': {30: 1},\n",
              " 'further': {0: 1, 25: 1, 30: 1},\n",
              " 'increase': {1: 1, 44: 1, 38: 1},\n",
              " 'transcripts': {33: 1},\n",
              " 'Transient': {0: 1},\n",
              " 'several': {0: 1},\n",
              " 'promoter-proximal': {0: 1},\n",
              " 'contribute': {44: 1, 38: 1},\n",
              " 'most': {41: 1, 8: 1, 0: 1},\n",
              " 'importantly': {30: 1},\n",
              " 'hand': {1: 1},\n",
              " 'depends': {12: 1},\n",
              " '78-nucleotide': {0: 1},\n",
              " 'segment': {1: 1},\n",
              " '1.3': {42: 1},\n",
              " 'kilobases': {33: 1},\n",
              " 'start': {1: 1, 38: 1},\n",
              " 'IL-2-inducible': {0: 1},\n",
              " 'lies': {12: 1},\n",
              " 'becomes': {12: 1},\n",
              " 'DNase': {1: 1},\n",
              " 'hypersensitive': {0: 1},\n",
              " 'normal': {0: 1},\n",
              " 'induced': {19: 1, 0: 1, 35: 1},\n",
              " 'three': {42: 1},\n",
              " 'Two': {42: 1},\n",
              " 'STAT': {33: 1, 1: 1},\n",
              " 'Hematopoietic': {0: 1},\n",
              " 'lineage': {1: 1},\n",
              " 'commitment': {1: 1},\n",
              " 'factors': {33: 1, 1: 1},\n",
              " 'review': {1: 1, 44: 1, 38: 1},\n",
              " 'focuses': {12: 1},\n",
              " 'roles': {33: 1},\n",
              " 'hematopoietic': {0: 1},\n",
              " 'A': {1: 1, 24: 1, 3: 1},\n",
              " 'brief': {0: 1},\n",
              " 'introduction': {1: 1},\n",
              " 'asymmetric': {0: 1},\n",
              " 'division': {1: 1},\n",
              " 'discussion': {1: 1},\n",
              " 'methods': {33: 1},\n",
              " 'important': {0: 1},\n",
              " 'specifying': {4: 1},\n",
              " 'types': {33: 1},\n",
              " 'Next': {30: 1},\n",
              " 'presented': {19: 1, 35: 1},\n",
              " 'use': {1: 1, 44: 1, 38: 1},\n",
              " 'embryonic': {0: 1},\n",
              " 'stem': {1: 1, 38: 1},\n",
              " 'analysis': {1: 1},\n",
              " 'targeted': {19: 1, 35: 1},\n",
              " 'disruption': {1: 1},\n",
              " 'analyze': {44: 1, 38: 1},\n",
              " 'hematopoiesis': {1: 1},\n",
              " 'Finally': {30: 1},\n",
              " 'status': {1: 1},\n",
              " 'our': {22: 1},\n",
              " 'current': {0: 1, 1: 1},\n",
              " 'knowledge': {1: 1},\n",
              " 'concerning': {4: 1},\n",
              " 'myeloid': {0: 1, 1: 1},\n",
              " 'lymphoid': {0: 1},\n",
              " 'summarized': {35: 1},\n",
              " 'Epstein-Barr': {0: 1},\n",
              " 'replicative': {0: 1, 1: 1},\n",
              " 'de': {18: 1},\n",
              " 'novo': {18: 1},\n",
              " 'thymocytes': {33: 1},\n",
              " 'simultaneous': {0: 1},\n",
              " 'early': {0: 1, 30: 1},\n",
              " 'BZLF-1': {1: 1},\n",
              " 'its': {22: 1},\n",
              " 'repressor': {1: 1},\n",
              " 'RAZ': {1: 1},\n",
              " 'EBV': {0: 1, 1: 1, 33: 1},\n",
              " 'known': {0: 1, 35: 1},\n",
              " 'infect': {44: 1},\n",
              " 'epithelial': {0: 1, 1: 1},\n",
              " 'others': {33: 1},\n",
              " 'can': {43: 1},\n",
              " 'subset': {1: 1},\n",
              " 'Infection': {1: 1},\n",
              " 'accompanied': {19: 1, 35: 1},\n",
              " 'linear': {0: 1, 1: 1},\n",
              " 'genome': {1: 1},\n",
              " '8': {42: 1},\n",
              " 'hr': {1: 1},\n",
              " 'Circularization': {1: 1},\n",
              " 'detected': {19: 1, 0: 1, 35: 1},\n",
              " 'contrast': {1: 1},\n",
              " 'where': {26: 1},\n",
              " 'circularize': {44: 1},\n",
              " '24': {42: 1},\n",
              " 'BamHI': {1: 1},\n",
              " 'ZLF-1': {1: 1},\n",
              " 'product': {1: 1},\n",
              " 'ZEBRA': {1: 1},\n",
              " 'RT-PCR': {1: 1},\n",
              " 'observed': {19: 1, 35: 1},\n",
              " 'fusion': {1: 1},\n",
              " 'transcript': {1: 1},\n",
              " 'comprised': {19: 1, 35: 1},\n",
              " 'locus': {1: 1},\n",
              " 'BRLF-1': {1: 1},\n",
              " 'immunoprecipitation': {1: 1},\n",
              " 'addition': {1: 1},\n",
              " 'EBNA-1': {1: 1},\n",
              " 'transcribed': {19: 1, 35: 1},\n",
              " 'Fp': {1: 1},\n",
              " 'rather': {30: 1},\n",
              " 'than': {23: 1},\n",
              " 'Cp/Wp': {1: 1},\n",
              " 'latently': {30: 1},\n",
              " 'Transcripts': {33: 1},\n",
              " 'encoding': {0: 1, 1: 1, 4: 1},\n",
              " 'gp350/220': {1: 1},\n",
              " 'coat': {1: 1},\n",
              " 'did': {19: 1},\n",
              " 'find': {44: 1, 38: 1},\n",
              " 'any': {24: 1},\n",
              " 'evidence': {1: 1},\n",
              " 'LMP-2A': {1: 1},\n",
              " 'EBER-1': {1: 1},\n",
              " 'loci': {33: 1},\n",
              " 'observations': {33: 1},\n",
              " 'differs': {12: 1},\n",
              " 'main': {0: 1},\n",
              " 'difference': {1: 1},\n",
              " 'ever': {30: 1},\n",
              " 'circularizes': {12: 1},\n",
              " 'Rather': {30: 1},\n",
              " 'remains': {12: 1},\n",
              " 'configuration': {1: 1},\n",
              " 'Identification': {1: 1},\n",
              " 'purification': {1: 1},\n",
              " 'Stat': {33: 1, 1: 1},\n",
              " 'activated': {19: 1, 0: 1, 35: 1},\n",
              " 'key': {0: 1},\n",
              " 'cytokine': {1: 1},\n",
              " 'immune': {0: 1},\n",
              " 'Following': {4: 1},\n",
              " 'various': {0: 1},\n",
              " 'chains': {33: 1},\n",
              " 'transcriptionally': {30: 1},\n",
              " 'turn': {1: 1},\n",
              " 'secreted': {19: 1, 0: 1, 35: 1},\n",
              " 'serves': {12: 1},\n",
              " 'differentiation': {1: 1},\n",
              " 'Several': {0: 1},\n",
              " 'recent': {0: 1},\n",
              " 'studies': {33: 1, 12: 1},\n",
              " 'implicated': {19: 1, 35: 1},\n",
              " 'Jak': {1: 1},\n",
              " 'lead': {19: 1, 1: 1, 35: 1, 44: 1, 38: 1},\n",
              " 'set': {19: 1, 1: 1, 38: 1, 35: 1},\n",
              " 'out': {23: 1, 30: 1, 40: 1},\n",
              " 'contain': {44: 1, 38: 1},\n",
              " 'activities': {33: 1},\n",
              " 'Similar': {0: 1},\n",
              " 'transformed': {0: 1, 19: 1, 35: 1},\n",
              " 'line': {1: 1},\n",
              " 'termed': {19: 1, 35: 1},\n",
              " 'YT': {1: 1},\n",
              " 'purified': {0: 1, 19: 1, 35: 1},\n",
              " 'principal': {0: 1},\n",
              " 'component': {1: 1},\n",
              " 'bears': {12: 1},\n",
              " 'relatedness': {1: 1},\n",
              " 'prolactin-induced': {0: 1},\n",
              " 'sheep': {1: 1},\n",
              " 'mammary': {0: 1},\n",
              " 'gland': {1: 1},\n",
              " 'tissue': {1: 1},\n",
              " 'hypothesize': {38: 1},\n",
              " 'hStat5': {1: 1},\n",
              " 'helps': {12: 1},\n",
              " 'govern': {44: 1, 38: 1},\n",
              " 'biological': {0: 1},\n",
              " 'E2F-1': {1: 1},\n",
              " 'cyclin-like': {0: 1},\n",
              " 'repair': {1: 1},\n",
              " 'enzyme': {1: 1},\n",
              " 'uracil-DNA': {1: 1},\n",
              " 'glycosylase': {1: 1},\n",
              " 'autoregulatory': {0: 1, 1: 1},\n",
              " 'mechanism': {1: 1},\n",
              " 'cycle-dependent': {0: 1},\n",
              " 'regulates': {12: 1},\n",
              " 'species': {33: 1},\n",
              " 'UDG': {1: 1},\n",
              " 'osteosarcoma': {1: 1},\n",
              " 'Saos-2': {1: 1},\n",
              " 'deletion': {1: 1},\n",
              " 'sequences': {33: 1, 12: 1},\n",
              " 'activates': {12: 1},\n",
              " 'E2F': {1: 1},\n",
              " 'putative': {0: 1},\n",
              " 'downstream': {0: 1, 30: 1},\n",
              " 'located': {0: 1},\n",
              " 'exon': {1: 1},\n",
              " 'E2F-1/DP1': {1: 1},\n",
              " 'functional': {0: 1},\n",
              " 'relationship': {1: 1},\n",
              " 'between': {23: 1},\n",
              " 'High': {0: 1},\n",
              " 'assay': {1: 1, 44: 1},\n",
              " 'result': {1: 1, 44: 1, 38: 1},\n",
              " 'E2F-mediated': {0: 1},\n",
              " 'Overexpression': {1: 1},\n",
              " 'Saos': {1: 1},\n",
              " 'growth': {1: 1},\n",
              " 'late': {0: 1, 30: 1},\n",
              " 'G1': {1: 1},\n",
              " 'transiently': {30: 1},\n",
              " 'arrest': {1: 1, 44: 1},\n",
              " 'progressing': {4: 1},\n",
              " 'S': {1: 1, 3: 1},\n",
              " 'hypothetical': {0: 1},\n",
              " 'integrates': {12: 1},\n",
              " 'one': {1: 1, 42: 1, 29: 1},\n",
              " 'cycle': {1: 1},\n",
              " 'likely': {0: 1, 30: 1},\n",
              " 'implicates': {12: 1},\n",
              " 'multifunctional': {0: 1},\n",
              " 'enzymes': {33: 1},\n",
              " 'possibly': {30: 1},\n",
              " 'damage': {1: 1},\n",
              " 'negative': {0: 1},\n",
              " 'Cellular': {0: 1},\n",
              " 'molecular': {0: 1, 1: 1},\n",
              " 'IFN-gamma': {2: 1, 1: 1},\n",
              " 'IL-12': {1: 1},\n",
              " 'Interferon-gamma': {1: 1},\n",
              " 'immunoregulatory': {0: 1},\n",
              " 'produced': {19: 1, 35: 1},\n",
              " 'predominantly': {30: 1},\n",
              " 'large': {0: 1},\n",
              " 'granular': {0: 1},\n",
              " 'LGL': {1: 1},\n",
              " 'extracellular': {0: 1},\n",
              " 'particular': {0: 1},\n",
              " 'interleukins': {33: 1},\n",
              " 'ILs': {33: 1},\n",
              " 'potent': {0: 1},\n",
              " 'inducers': {33: 1},\n",
              " 'Although': {23: 1},\n",
              " 'it': {29: 1},\n",
              " 'reported': {19: 1, 35: 1},\n",
              " 'there': {34: 1, 30: 1},\n",
              " 'some': {24: 1},\n",
              " 'produce': {44: 1, 38: 1},\n",
              " 'yet': {30: 1, 36: 1},\n",
              " 'report': {1: 1, 38: 1},\n",
              " 'killer': {1: 1},\n",
              " 'responds': {12: 1},\n",
              " 'similar': {0: 1},\n",
              " 'present': {0: 1, 1: 1, 44: 1, 38: 1},\n",
              " 'NK3.3': {1: 1},\n",
              " 'derived': {0: 1, 19: 1, 35: 1},\n",
              " 'measured': {19: 1, 35: 1},\n",
              " 'increases': {33: 1, 12: 1},\n",
              " 'granulocyte-macrophage': {0: 1, 1: 1},\n",
              " 'colony-stimulating': {0: 1},\n",
              " 'GM-CSF': {1: 1},\n",
              " 'cytoplasmic': {0: 1},\n",
              " 'when': {26: 1},\n",
              " 'together': {30: 1},\n",
              " 'synergized': {19: 1},\n",
              " 'induction': {1: 1},\n",
              " 'synergy': {1: 1},\n",
              " 'attributed': {35: 1},\n",
              " 'increased': {19: 1, 35: 1},\n",
              " 'accumulation': {1: 1},\n",
              " 'stability': {1: 1},\n",
              " 'mRNAs': {33: 1},\n",
              " 'investigate': {44: 1, 38: 1},\n",
              " 'five': {42: 1},\n",
              " 'cyclosporin': {1: 1},\n",
              " 'CsA': {1: 1},\n",
              " 'transforming': {0: 1, 1: 1, 4: 1},\n",
              " 'factor-beta': {1: 1},\n",
              " 'cycloheximide': {1: 1},\n",
              " 'genistein': {1: 1},\n",
              " 'staurosporine': {1: 1},\n",
              " 'analyzing': {4: 1},\n",
              " 'new': {0: 1},\n",
              " 'synthesis': {1: 1},\n",
              " 'appears': {12: 1},\n",
              " 'only': {0: 1, 30: 1},\n",
              " 'partially': {30: 1},\n",
              " 'depend': {44: 1, 38: 1},\n",
              " 'Furthermore': {30: 1},\n",
              " 'inhibitor': {1: 1},\n",
              " 'generally': {30: 1},\n",
              " 'inactive': {0: 1},\n",
              " 'suppression': {1: 1},\n",
              " 'agents': {33: 1},\n",
              " 'independent': {0: 1},\n",
              " 'inhibition': {1: 1},\n",
              " 'AP1': {1: 1},\n",
              " 'complexes': {33: 1},\n",
              " 'leading': {4: 1},\n",
              " 'serve': {44: 1, 38: 1},\n",
              " 'dissecting': {4: 1},\n",
              " 'biochemical': {0: 1},\n",
              " 'events': {33: 1},\n",
              " 'p95vav': {1: 1},\n",
              " 'Stimulation': {1: 1},\n",
              " 'TCR': {1: 1},\n",
              " 'multiple': {0: 1},\n",
              " 'numerous': {0: 1},\n",
              " 'substrates': {33: 1},\n",
              " 'One': {42: 1},\n",
              " 'substrate': {0: 1, 1: 1},\n",
              " 'expressed': {19: 1, 35: 1},\n",
              " 'exclusively': {30: 1},\n",
              " 'trophoblast': {1: 1},\n",
              " 'contains': {12: 1},\n",
              " 'structural': {0: 1},\n",
              " 'motifs': {33: 1},\n",
              " 'Src': {1: 1},\n",
              " 'homology': {1: 1},\n",
              " 'pleckstrin': {1: 1},\n",
              " 'domains': {33: 1},\n",
              " 'guanine': {0: 1, 1: 1},\n",
              " 'nucleotide': {0: 1, 1: 1},\n",
              " 'exchange': {1: 1},\n",
              " 'TCR-mediated': {0: 1},\n",
              " 'processes': {33: 1, 12: 1},\n",
              " 'unclear': {0: 1},\n",
              " 'overexpression': {1: 1},\n",
              " 'alone': {30: 1},\n",
              " 'Jurkat': {1: 1},\n",
              " 'NFAT': {1: 1},\n",
              " 'synergizes': {12: 1},\n",
              " 'inducing': {1: 1, 4: 1},\n",
              " 'NFAT-': {1: 1},\n",
              " 'interleukin-2-dependent': {0: 1},\n",
              " 'G-protein-coupled': {0: 1},\n",
              " 'modulated': {0: 1, 19: 1, 35: 1},\n",
              " 'suggesting': {4: 1},\n",
              " 'removal': {1: 1},\n",
              " '67': {42: 1},\n",
              " 'amino': {1: 1},\n",
              " 'acids': {33: 1},\n",
              " 'NIH': {1: 1},\n",
              " '3T3': {1: 1},\n",
              " 'p95vav-induced': {0: 1},\n",
              " 'mimicked': {19: 1, 35: 1},\n",
              " 'though': {23: 1, 30: 1},\n",
              " 'upon': {23: 1},\n",
              " 'Raf': {1: 1},\n",
              " 'FK506': {1: 1},\n",
              " 'calcineurin': {1: 1},\n",
              " 'dissect': {44: 1},\n",
              " 'involvement': {1: 1},\n",
              " 'analyzed': {19: 1, 35: 1},\n",
              " 'mutants': {33: 1},\n",
              " 'deficient': {0: 1, 1: 1},\n",
              " 'showed': {19: 1, 35: 1},\n",
              " 'However': {30: 1, 26: 1},\n",
              " 'does': {12: 1},\n",
              " 'appear': {44: 1, 38: 1},\n",
              " 'influence': {1: 1, 44: 1, 38: 1},\n",
              " 'TCR-induced': {0: 1},\n",
              " 'free': {0: 1},\n",
              " 'Taken': {35: 1},\n",
              " 'plays': {12: 1},\n",
              " 'unidentified': {0: 1},\n",
              " 'proximal': {0: 1, 1: 1},\n",
              " 'position': {1: 1},\n",
              " 'Positive': {0: 1},\n",
              " 'AML1-related': {0: 1},\n",
              " 'PEBP2': {1: 1},\n",
              " 'consensus': {1: 1},\n",
              " 'polyomavirus': {1: 1},\n",
              " 'binding-protein': {1: 1},\n",
              " 'consists': {12: 1},\n",
              " 'beta': {1: 1},\n",
              " 'subunits': {33: 1},\n",
              " 'There': {34: 1},\n",
              " 'subunit': {1: 1},\n",
              " 'homologue': {1: 1},\n",
              " 'AML1': {1: 1},\n",
              " 'breakpoints': {33: 1},\n",
              " 't': {1: 1},\n",
              " ';': {45: 1},\n",
              " '21': {42: 1},\n",
              " 'leukemias': {33: 1, 1: 1},\n",
              " 'A1': {1: 1},\n",
              " 'A-gene': {0: 1},\n",
              " 'B1': {1: 1},\n",
              " 'B2': {1: 1},\n",
              " 'B-encoded': {0: 1},\n",
              " 'isomers': {33: 1},\n",
              " 'bound': {19: 1, 0: 1, 35: 1},\n",
              " 'promoter-driven': {0: 1},\n",
              " 'reporter': {1: 1},\n",
              " 'plasmid': {1: 1},\n",
              " 'unstimulated': {0: 1},\n",
              " '12-O-tetradecanoylphorbol': {1: 1},\n",
              " '13-acetate/phytohemagglutinin-stimulated': {0: 1},\n",
              " 'suppressed': {19: 1, 35: 1},\n",
              " 'Coexpression': {1: 1},\n",
              " 'determined': {19: 1, 35: 1},\n",
              " 'B1/alpha': {1: 1},\n",
              " 'ratio': {1: 1},\n",
              " 'extract': {1: 1},\n",
              " 'contained': {19: 1, 35: 1},\n",
              " 'site-binding': {0: 1},\n",
              " 'cross-reacted': {19: 1},\n",
              " 'antimouse': {0: 1},\n",
              " 'Northern': {0: 1, 1: 1},\n",
              " 'blot': {1: 1},\n",
              " 'indicated': {19: 1, 35: 1},\n",
              " 'determine': {44: 1, 38: 1, 35: 1},\n",
              " 'precise': {0: 1},\n",
              " 'suggested': {19: 1, 35: 1},\n",
              " 'importance': {1: 1},\n",
              " 'relative': {0: 1, 1: 1},\n",
              " 'isoforms': {33: 1, 12: 1},\n",
              " 'regulating': {1: 1, 4: 1},\n",
              " 'priming': {0: 1, 1: 1, 4: 1},\n",
              " 'enhances': {12: 1},\n",
              " 'LPS-induced': {0: 1},\n",
              " 'TNF': {1: 1},\n",
              " 'augmenting': {4: 1},\n",
              " 'MRNA': {1: 1},\n",
              " '/macrophages': {33: 1},\n",
              " 'bacterial': {0: 1},\n",
              " 'endotoxin': {1: 1},\n",
              " 'lipopolysaccharide': {1: 1},\n",
              " 'critical': {0: 1},\n",
              " 'highly': {30: 1},\n",
              " 'defence': {1: 1},\n",
              " 'augmentation': {1: 1},\n",
              " 'LPS': {33: 1, 1: 1},\n",
              " 'responses': {33: 1},\n",
              " 'interferon': {1: 1},\n",
              " 'gamma': {1: 1, 44: 1},\n",
              " 'referred': {35: 1},\n",
              " 'established': {0: 1, 19: 1, 35: 1},\n",
              " 'poorly': {30: 1},\n",
              " 'Using': {4: 1},\n",
              " 'tumour': {0: 1, 1: 1},\n",
              " 'necrosis': {1: 1},\n",
              " 'designed': {35: 1},\n",
              " 'analyse': {44: 1, 38: 1},\n",
              " 'detail': {1: 1},\n",
              " 'Priming': {1: 1, 4: 1},\n",
              " 'manifested': {35: 1},\n",
              " 'level': {1: 1},\n",
              " 'pre-treatment': {1: 1},\n",
              " 'affected': {19: 1, 0: 1, 35: 1},\n",
              " 'magnitude': {1: 1},\n",
              " 'sensitivity': {1: 1},\n",
              " 'occurred': {19: 1, 35: 1},\n",
              " 'after': {23: 1, 30: 1},\n",
              " 'hours': {33: 1},\n",
              " 'treatment': {1: 1},\n",
              " 'primed': {0: 1, 19: 1, 35: 1},\n",
              " 'state': {1: 1},\n",
              " 'M-CSF': {1: 1},\n",
              " 'Primed': {35: 1},\n",
              " 'higher': {27: 1},\n",
              " 'rate': {1: 1},\n",
              " 'freshly': {30: 1},\n",
              " 'isolated': {19: 1, 35: 1},\n",
              " 'marked': {0: 1, 35: 1},\n",
              " 'factor-kappa': {1: 1},\n",
              " 'oligonucleotide': {0: 1, 1: 1},\n",
              " 'An': {24: 1},\n",
              " 'finding': {1: 1, 4: 1},\n",
              " 'much': {0: 1, 30: 1},\n",
              " 'more': {27: 1, 25: 1},\n",
              " 'unprimed': {0: 1},\n",
              " 'T1/2': {1: 1},\n",
              " '6-8-fold': {30: 1},\n",
              " 'duration': {1: 1},\n",
              " 'longer': {27: 1, 25: 1, 30: 1},\n",
              " 'being': {4: 1},\n",
              " 'greater': {27: 1},\n",
              " 'possessed': {19: 1, 35: 1},\n",
              " 'H-89': {1: 1},\n",
              " 'substantially': {30: 1},\n",
              " 'had': {19: 1},\n",
              " 'Myc-associated': {0: 1},\n",
              " 'zinc': {1: 1},\n",
              " 'developmental': {0: 1},\n",
              " 'minimal': {0: 1},\n",
              " 'mutagenesis': {1: 1},\n",
              " 'analyses': {33: 1},\n",
              " 'initiator-like': {0: 1},\n",
              " 'cap': {1: 1},\n",
              " 'Ets': {33: 1, 1: 1},\n",
              " 'MAZ': {1: 1},\n",
              " 'predominant': {0: 1},\n",
              " 'last': {0: 1},\n",
              " 'resembles': {12: 1},\n",
              " 'ME1a1': {1: 1},\n",
              " 'G3AG4AG3': {1: 1},\n",
              " 'motif': {1: 1},\n",
              " ...}"
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ],
      "source": [
        "# adding features\n",
        "word_features = {word: {pos_indices[pos]: 1 for pos in pos_set} for (word, pos_set) in words_pos_dict.items()}\n",
        "word_features"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4ZJKO2vx4V5c"
      },
      "source": [
        "### Structural Features\n",
        "\n",
        "The paper uses a set of 22 structural features for the words (i.e., word contains certain characters, contains numbers, etc.). We implement the rules for each of these features and compute them for each word. Each feature has a unique index assigned to it (*starting from 48* since the last POS feature has index 47) which will be used to assign the feature to the word in the feature dict."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "nimZpB2xrP5n"
      },
      "outputs": [],
      "source": [
        "def capital_letter_indices(s: str) -> List[int]:\n",
        "  return [idx for idx in range(len(s)) if s[idx].isupper()]\n",
        "\n",
        "def str_contains_digit(s: str) -> bool:\n",
        "  return any(ch.isdigit() for ch in s)\n",
        "\n",
        "def str_contains_letter(s: str) -> bool:\n",
        "  return any(ch.isalpha() for ch in s)\n",
        "\n",
        "def all_capital(s: str) -> bool:\n",
        "  # feat 11\n",
        "  return s.isalpha() and s.isupper()\n",
        "\n",
        "def all_lower(s: str) -> bool:\n",
        "  # feat 20\n",
        "  return s.isalpha() and s.islower()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "FrsyUyALrv67"
      },
      "outputs": [],
      "source": [
        "# testing rule functions\n",
        "assert all_capital('aaaa') == False\n",
        "assert all_capital('AhEllO') == False\n",
        "assert all_capital('HELLO') == True\n",
        "assert all_capital('HELLO3') == False\n",
        "\n",
        "assert capital_letter_indices(\"hELlo\") == [1, 2]\n",
        "assert capital_letter_indices('hello') == []\n",
        "\n",
        "assert str_contains_digit('b5b') == True\n",
        "assert str_contains_digit('hello') == False\n",
        "\n",
        "assert str_contains_letter('b5b') == True\n",
        "assert str_contains_letter('hello') == True\n",
        "assert str_contains_letter('12345') == False"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "V1RcgVaSw5JM"
      },
      "outputs": [],
      "source": [
        "num_structural_feat = 22\n",
        "\n",
        "# dict containing functions that return true if given string has the i-th feature\n",
        "# indices for features matches those given in paper\n",
        "structural_rules = {\n",
        "    1: lambda s: s.isdigit(),\n",
        "    2: lambda s: s.count('/') == 1,\n",
        "    3: lambda s: s.count('/') == 2,\n",
        "    4: lambda s: '$' in s,\n",
        "    5: lambda s: '%' in s,\n",
        "    6: lambda s: ',' in s,\n",
        "    7: lambda s: '.' in s,\n",
        "    8: lambda s: ':' in s,\n",
        "    9: lambda s: '-' in s,\n",
        "    10: lambda s: str_contains_letter(s) and str_contains_digit(s) and s.count('/') > 0,\n",
        "    11: all_capital,\n",
        "    12: lambda s: 0 in capital_letter_indices(s) and '.' in s,\n",
        "    13: lambda s: len(capital_letter_indices(s)) > 0 and '.' in s,\n",
        "    14: lambda s: str_contains_letter(s) and '$' in s,\n",
        "    15: lambda s: str_contains_letter(s) and '.' in s,\n",
        "    16: lambda s: len(capital_letter_indices(s)) > 0,\n",
        "    17: lambda s: str_contains_letter(s) and str_contains_digit(s),\n",
        "    18: lambda s: 0 in capital_letter_indices(s), # first letter capital\n",
        "    19: lambda s: any(idx > 0 and idx < len(s)-1 for idx in capital_letter_indices(s)), # capital letter in middle of word\n",
        "    20: all_lower,\n",
        "    21: lambda s: str_contains_letter(s) and '-' in s\n",
        "}\n",
        "# 22nd feature assigned to any word that doesn't have any of the other 21 features"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "id": "USAu4RSu1joi"
      },
      "outputs": [],
      "source": [
        "for word in word_features.keys():\n",
        "  word_features[word].update({num_pos_feat + (idx - 1) : 1 for (idx, f) in structural_rules.items() if f(word)})\n",
        "  if max(word_features[word].keys()) <= 47: # if word has none of the other structural features\n",
        "    word_features[word].update({num_pos_feat + (22 - 1): 1})"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "h8mHLMpHWK5W",
        "outputId": "1baa8130-ef66-440e-c425-fd1352e0ca52"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'IL-2': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'gene': {1: 1, 67: 1},\n",
              " 'expression': {1: 1, 67: 1},\n",
              " 'and': {36: 1, 67: 1},\n",
              " 'NF-kappa': {1: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'B': {1: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'activation': {1: 1, 67: 1},\n",
              " 'through': {23: 1, 67: 1},\n",
              " 'CD28': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'requires': {12: 1, 67: 1},\n",
              " 'reactive': {0: 1, 67: 1},\n",
              " 'oxygen': {1: 1, 67: 1},\n",
              " 'production': {1: 1, 67: 1},\n",
              " 'by': {23: 1, 67: 1},\n",
              " '5-lipoxygenase': {0: 1, 1: 1, 56: 1, 64: 1, 68: 1},\n",
              " '.': {37: 1, 54: 1},\n",
              " 'Activation': {1: 1, 63: 1, 65: 1},\n",
              " 'of': {23: 1, 67: 1},\n",
              " 'the': {24: 1, 13: 1, 67: 1},\n",
              " 'surface': {1: 1, 67: 1},\n",
              " 'receptor': {1: 1, 67: 1},\n",
              " 'provides': {12: 1, 67: 1},\n",
              " 'a': {1: 1, 20: 1, 4: 1, 24: 1, 46: 1, 13: 1, 67: 1},\n",
              " 'major': {0: 1, 67: 1},\n",
              " 'costimulatory': {0: 1, 1: 1, 67: 1},\n",
              " 'signal': {0: 1, 1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'for': {23: 1, 67: 1},\n",
              " 'T': {0: 1, 1: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'cell': {1: 1, 67: 1},\n",
              " 'resulting': {4: 1, 67: 1},\n",
              " 'in': {18: 1, 1: 1, 23: 1, 30: 1, 40: 1, 67: 1},\n",
              " 'enhanced': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'interleukin-2': {1: 1, 56: 1, 64: 1, 68: 1},\n",
              " '(': {32: 1, 69: 1},\n",
              " ')': {2: 1, 47: 1, 69: 1},\n",
              " 'proliferation': {1: 1, 67: 1},\n",
              " 'In': {18: 1, 23: 1, 63: 1, 65: 1},\n",
              " 'primary': {0: 1, 67: 1},\n",
              " 'lymphocytes': {33: 1, 67: 1},\n",
              " 'we': {29: 1, 67: 1},\n",
              " 'show': {44: 1, 38: 1, 67: 1},\n",
              " 'that': {23: 1, 24: 1, 11: 1, 67: 1},\n",
              " 'ligation': {1: 1, 67: 1},\n",
              " 'leads': {12: 1, 67: 1},\n",
              " 'to': {5: 1, 67: 1},\n",
              " 'rapid': {0: 1, 67: 1},\n",
              " 'intracellular': {0: 1, 67: 1},\n",
              " 'formation': {1: 1, 67: 1},\n",
              " 'intermediates': {33: 1, 67: 1},\n",
              " 'ROIs': {33: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'which': {11: 1, 17: 1, 67: 1},\n",
              " 'are': {38: 1, 67: 1},\n",
              " 'required': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'CD28-mediated': {0: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'B/CD28-responsive': {0: 1,\n",
              "  49: 1,\n",
              "  56: 1,\n",
              "  57: 1,\n",
              "  63: 1,\n",
              "  64: 1,\n",
              "  65: 1,\n",
              "  66: 1,\n",
              "  68: 1},\n",
              " 'complex': {0: 1, 1: 1, 67: 1},\n",
              " 'Delineation': {1: 1, 63: 1, 65: 1},\n",
              " 'signaling': {0: 1, 1: 1, 4: 1, 67: 1},\n",
              " 'cascade': {1: 1, 67: 1},\n",
              " 'was': {19: 1, 67: 1},\n",
              " 'found': {19: 1, 35: 1, 67: 1},\n",
              " 'involve': {44: 1, 38: 1, 67: 1},\n",
              " 'protein': {1: 1, 67: 1},\n",
              " 'tyrosine': {1: 1, 67: 1},\n",
              " 'kinase': {1: 1, 67: 1},\n",
              " 'activity': {1: 1, 67: 1},\n",
              " ',': {37: 1, 21: 1, 53: 1},\n",
              " 'followed': {19: 1, 35: 1, 67: 1},\n",
              " 'phospholipase': {1: 1, 67: 1},\n",
              " 'A2': {1: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'Our': {22: 1, 63: 1, 65: 1},\n",
              " 'data': {33: 1, 1: 1, 67: 1},\n",
              " 'suggest': {44: 1, 38: 1, 67: 1},\n",
              " 'lipoxygenase': {1: 1, 67: 1},\n",
              " 'metabolites': {33: 1, 67: 1},\n",
              " 'activate': {44: 1, 38: 1, 67: 1},\n",
              " 'ROI': {33: 1, 1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'then': {30: 1, 67: 1},\n",
              " 'induce': {44: 1, 38: 1, 67: 1},\n",
              " 'via': {23: 1, 67: 1},\n",
              " 'These': {24: 1, 63: 1, 65: 1},\n",
              " 'findings': {33: 1, 67: 1},\n",
              " 'should': {43: 1, 67: 1},\n",
              " 'be': {44: 1, 67: 1},\n",
              " 'useful': {0: 1, 67: 1},\n",
              " 'therapeutic': {0: 1, 67: 1},\n",
              " 'strategies': {33: 1, 67: 1},\n",
              " 'development': {1: 1, 67: 1},\n",
              " 'immunosuppressants': {33: 1, 67: 1},\n",
              " 'targeting': {1: 1, 4: 1, 67: 1},\n",
              " 'pathway': {1: 1, 67: 1},\n",
              " 'The': {24: 1, 3: 1, 63: 1, 65: 1},\n",
              " 'peri-kappa': {1: 1, 56: 1, 68: 1},\n",
              " 'site': {1: 1, 67: 1},\n",
              " 'mediates': {12: 1, 67: 1},\n",
              " 'human': {0: 1, 1: 1, 67: 1},\n",
              " 'immunodeficiency': {1: 1, 67: 1},\n",
              " 'virus': {1: 1, 67: 1},\n",
              " 'type': {1: 1, 67: 1},\n",
              " '2': {42: 1, 46: 1, 48: 1},\n",
              " 'enhancer': {1: 1, 67: 1},\n",
              " 'monocytes': {33: 1, 67: 1},\n",
              " 'but': {36: 1, 67: 1},\n",
              " 'not': {30: 1, 67: 1},\n",
              " 'cells': {33: 1, 67: 1},\n",
              " 'Human': {0: 1, 63: 1, 65: 1},\n",
              " 'HIV-2': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'like': {0: 1, 23: 1, 67: 1},\n",
              " 'HIV-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'causes': {33: 1, 12: 1, 67: 1},\n",
              " 'AIDS': {1: 1, 3: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'is': {19: 1, 12: 1, 67: 1},\n",
              " 'associated': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'with': {23: 1, 67: 1},\n",
              " 'cases': {33: 1, 67: 1},\n",
              " 'primarily': {30: 1, 67: 1},\n",
              " 'West': {0: 1, 3: 1, 63: 1, 65: 1},\n",
              " 'Africa': {3: 1, 63: 1, 65: 1},\n",
              " 'display': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'significant': {0: 1, 67: 1},\n",
              " 'differences': {33: 1, 67: 1},\n",
              " 'nucleic': {0: 1, 67: 1},\n",
              " 'acid': {0: 1, 1: 1, 67: 1},\n",
              " 'sequence': {1: 1, 67: 1},\n",
              " 'natural': {0: 1, 67: 1},\n",
              " 'history': {1: 1, 67: 1},\n",
              " 'clinical': {0: 1, 67: 1},\n",
              " 'disease': {1: 1, 67: 1},\n",
              " 'Consistent': {0: 1, 63: 1, 65: 1},\n",
              " 'these': {24: 1, 67: 1},\n",
              " 'have': {44: 1, 38: 1, 67: 1},\n",
              " 'previously': {30: 1, 67: 1},\n",
              " 'demonstrated': {19: 1, 35: 1, 67: 1},\n",
              " 'enhancer/promoter': {0: 1, 1: 1, 49: 1},\n",
              " 'region': {1: 1, 67: 1},\n",
              " 'functions': {33: 1, 12: 1, 67: 1},\n",
              " 'quite': {30: 1, 67: 1},\n",
              " 'differently': {30: 1, 67: 1},\n",
              " 'from': {23: 1, 67: 1},\n",
              " 'Whereas': {23: 1, 63: 1, 65: 1},\n",
              " 'following': {0: 1, 1: 1, 4: 1, 67: 1},\n",
              " 'T-cell': {1: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'stimulation': {1: 1, 67: 1},\n",
              " 'mediated': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'largely': {30: 1, 67: 1},\n",
              " 'binding': {0: 1, 1: 1, 4: 1, 67: 1},\n",
              " 'transcription': {1: 1, 67: 1},\n",
              " 'factor': {1: 1, 23: 1, 67: 1},\n",
              " 'two': {42: 1, 67: 1},\n",
              " 'adjacent': {0: 1, 67: 1},\n",
              " 'kappa': {1: 1, 67: 1},\n",
              " 'sites': {33: 1, 67: 1},\n",
              " 'long': {0: 1, 30: 1, 67: 1},\n",
              " 'terminal': {0: 1, 1: 1, 67: 1},\n",
              " 'repeat': {1: 1, 67: 1},\n",
              " 'dependent': {0: 1, 67: 1},\n",
              " 'on': {23: 1, 67: 1},\n",
              " 'four': {42: 1, 67: 1},\n",
              " 'cis-acting': {0: 1, 1: 1, 4: 1, 56: 1, 68: 1},\n",
              " 'elements': {33: 1, 67: 1},\n",
              " ':': {45: 1, 55: 1},\n",
              " 'single': {0: 1, 67: 1},\n",
              " 'purine-rich': {0: 1, 56: 1, 68: 1},\n",
              " 'PuB1': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'PuB2': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'pets': {1: 1, 67: 1},\n",
              " 'We': {29: 1, 63: 1, 65: 1},\n",
              " 'now': {1: 1, 30: 1, 67: 1},\n",
              " 'identified': {19: 1, 35: 1, 67: 1},\n",
              " 'novel': {0: 1, 67: 1},\n",
              " 'element': {1: 1, 67: 1},\n",
              " 'within': {23: 1, 67: 1},\n",
              " 'immediately': {30: 1, 67: 1},\n",
              " 'upstream': {0: 1, 30: 1, 67: 1},\n",
              " 'designated': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'This': {24: 1, 13: 1, 63: 1, 65: 1},\n",
              " 'conserved': {0: 1, 19: 1, 35: 1, 67: 1},\n",
              " 'among': {23: 1, 67: 1},\n",
              " 'isolates': {33: 1, 67: 1},\n",
              " 'closely': {30: 1, 67: 1},\n",
              " 'related': {0: 1, 35: 1, 67: 1},\n",
              " 'simian': {0: 1, 67: 1},\n",
              " 'transfection': {1: 1, 67: 1},\n",
              " 'assays': {33: 1, 1: 1, 67: 1},\n",
              " 'this': {24: 1, 67: 1},\n",
              " 'mediate': {44: 1, 38: 1, 67: 1},\n",
              " 'monocytic': {0: 1, 67: 1},\n",
              " 'lines': {33: 1, 67: 1},\n",
              " 'first': {0: 1, 30: 1, 67: 1},\n",
              " 'description': {1: 1, 67: 1},\n",
              " 'an': {24: 1, 13: 1, 67: 1},\n",
              " 'displays': {12: 1, 67: 1},\n",
              " 'such': {0: 1, 23: 1, 15: 1, 67: 1},\n",
              " 'monocyte': {0: 1, 1: 1, 67: 1},\n",
              " 'specificity': {1: 1, 67: 1},\n",
              " 'no': {24: 1, 30: 1, 67: 1},\n",
              " 'comparable': {0: 1, 67: 1},\n",
              " 'has': {12: 1, 67: 1},\n",
              " 'been': {35: 1, 67: 1},\n",
              " 'clearly': {30: 1, 67: 1},\n",
              " 'defined': {19: 1, 35: 1, 67: 1},\n",
              " 'While': {23: 1, 63: 1, 65: 1},\n",
              " 'nuclear': {0: 1, 67: 1},\n",
              " 's': {33: 1, 1: 1, 12: 1, 67: 1},\n",
              " 'both': {24: 1, 30: 1, 36: 1, 15: 1, 67: 1},\n",
              " 'peripheral': {0: 1, 67: 1},\n",
              " 'blood': {1: 1, 67: 1},\n",
              " 'binds': {12: 1, 67: 1},\n",
              " 'electrophoretic': {0: 1, 67: 1},\n",
              " 'mobility': {1: 1, 67: 1},\n",
              " 'shift': {1: 1, 44: 1, 67: 1},\n",
              " 'either': {24: 1, 30: 1, 36: 1, 67: 1},\n",
              " 'different': {0: 1, 67: 1},\n",
              " 'versus': {36: 1, 67: 1},\n",
              " 'or': {36: 1, 67: 1},\n",
              " 'recognizing': {4: 1, 67: 1},\n",
              " 'undergoes': {12: 1, 67: 1},\n",
              " 'differential': {0: 1, 67: 1},\n",
              " 'modification': {1: 1, 67: 1},\n",
              " 'thus': {30: 1, 67: 1},\n",
              " 'supporting': {1: 1, 4: 1, 67: 1},\n",
              " 'Further': {0: 1, 23: 1, 30: 1, 63: 1, 65: 1},\n",
              " 'while': {23: 1, 67: 1},\n",
              " 'specific': {0: 1, 67: 1},\n",
              " 'constitutive': {0: 1, 67: 1},\n",
              " 'seen': {35: 1, 67: 1},\n",
              " 'phorbol': {1: 1, 67: 1},\n",
              " 'esters': {33: 1, 67: 1},\n",
              " 'induces': {12: 1, 67: 1},\n",
              " 'additional': {0: 1, 67: 1},\n",
              " 'Understanding': {4: 1, 63: 1, 65: 1},\n",
              " 'monocyte-specific': {0: 1, 56: 1, 68: 1},\n",
              " 'function': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'may': {1: 1, 43: 1, 67: 1},\n",
              " 'ultimately': {30: 1, 67: 1},\n",
              " 'provide': {44: 1, 38: 1, 67: 1},\n",
              " 'insight': {1: 1, 67: 1},\n",
              " 'into': {23: 1, 67: 1},\n",
              " 'role': {1: 1, 67: 1},\n",
              " 'play': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'HIV': {1: 1, 38: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'pathogenesis': {1: 1, 67: 1},\n",
              " 'E1A': {1: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'susceptibility': {1: 1, 67: 1},\n",
              " 'killing': {1: 1, 4: 1, 67: 1},\n",
              " 'NK': {1: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'immortalization': {1: 1, 67: 1},\n",
              " 'adenovirus': {1: 1, 67: 1},\n",
              " 'infection': {1: 1, 67: 1},\n",
              " 'Adenovirus': {1: 1, 63: 1, 65: 1},\n",
              " 'Ad': {1: 1, 63: 1, 65: 1},\n",
              " 'were': {19: 1, 67: 1},\n",
              " 'used': {19: 1, 35: 1, 67: 1},\n",
              " 'model': {0: 1, 1: 1, 44: 1, 67: 1},\n",
              " 'changes': {33: 1, 12: 1, 67: 1},\n",
              " 'caused': {19: 1, 35: 1, 67: 1},\n",
              " 'transient': {0: 1, 1: 1, 67: 1},\n",
              " 'vs': {36: 1, 67: 1},\n",
              " 'stable': {0: 1, 67: 1},\n",
              " 'Only': {30: 1, 63: 1, 65: 1},\n",
              " 'stably': {30: 1, 67: 1},\n",
              " 'transfected': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'target': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'exhibited': {19: 1, 35: 1, 67: 1},\n",
              " 'cytolytic': {0: 1, 67: 1},\n",
              " 'despite': {23: 1, 67: 1},\n",
              " 'equivalent': {0: 1, 1: 1, 67: 1},\n",
              " 'levels': {33: 1, 67: 1},\n",
              " 'proteins': {33: 1, 67: 1},\n",
              " 'Ad-infected': {0: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'targets': {33: 1, 12: 1, 67: 1},\n",
              " 'inability': {1: 1, 67: 1},\n",
              " 'products': {33: 1, 67: 1},\n",
              " 'during': {23: 1, 67: 1},\n",
              " 'explained': {35: 1, 67: 1},\n",
              " 'inhibitory': {0: 1, 67: 1},\n",
              " 'effect': {1: 1, 44: 1, 67: 1},\n",
              " 'viral': {0: 1, 67: 1},\n",
              " 'otherwise': {0: 1, 30: 1, 67: 1},\n",
              " 'susceptible': {0: 1, 67: 1},\n",
              " 'effects': {33: 1, 67: 1},\n",
              " 'class': {1: 1, 67: 1},\n",
              " 'I': {1: 1, 46: 1, 42: 1, 29: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'MHC': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'antigen': {0: 1, 1: 1, 67: 1},\n",
              " 'phenotypes': {33: 1, 67: 1},\n",
              " 'infected': {0: 1, 35: 1, 67: 1},\n",
              " 'suggests': {12: 1, 67: 1},\n",
              " 'effective': {0: 1, 67: 1},\n",
              " 'immunologic': {0: 1, 67: 1},\n",
              " 'barrier': {1: 1, 67: 1},\n",
              " 'against': {23: 1, 67: 1},\n",
              " 'vivo': {18: 1, 67: 1},\n",
              " 'survival': {0: 1, 1: 1, 67: 1},\n",
              " 'neoplastic': {0: 1, 67: 1},\n",
              " 'progression': {1: 1, 67: 1},\n",
              " 'E1A-immortalized': {0: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'emerge': {44: 1, 38: 1, 67: 1},\n",
              " 'reservoir': {1: 1, 67: 1},\n",
              " 'persistently': {30: 1, 67: 1},\n",
              " 'host': {1: 1, 67: 1},\n",
              " 'Distinct': {0: 1, 63: 1, 65: 1},\n",
              " 'properties': {33: 1, 67: 1},\n",
              " 'identify': {44: 1, 38: 1, 67: 1},\n",
              " 'functionally': {30: 1, 67: 1},\n",
              " 'CD4': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'epitopes': {33: 1, 67: 1},\n",
              " 'coreceptor': {1: 1, 67: 1},\n",
              " 'interacts': {12: 1, 67: 1},\n",
              " 'non-polymorphic': {0: 1, 56: 1, 68: 1},\n",
              " 'regions': {33: 1, 67: 1},\n",
              " 'histocompatibility': {1: 1, 67: 1},\n",
              " 'II': {42: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'molecules': {33: 1, 67: 1},\n",
              " 'antigen-presenting': {0: 1, 56: 1, 68: 1},\n",
              " 'contributes': {12: 1, 67: 1},\n",
              " 'investigated': {19: 1, 35: 1, 67: 1},\n",
              " 'triggering': {1: 1, 4: 1, 67: 1},\n",
              " 'activating': {0: 1, 1: 1, 4: 1, 67: 1},\n",
              " 'signals': {33: 1, 12: 1, 67: 1},\n",
              " 'lymphoma': {1: 1, 67: 1},\n",
              " 'using': {4: 1, 67: 1},\n",
              " 'monoclonal': {0: 1, 67: 1},\n",
              " 'antibodies': {33: 1, 67: 1},\n",
              " 'mAb': {1: 1, 63: 1, 66: 1},\n",
              " 'recognize': {44: 1, 38: 1, 67: 1},\n",
              " 'demonstrate': {44: 1, 38: 1, 67: 1},\n",
              " 'delivers': {12: 1, 67: 1},\n",
              " 'capable': {0: 1, 67: 1},\n",
              " 'NF-AT': {1: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'anti-CD4': {0: 1, 56: 1, 63: 1, 64: 1, 66: 1, 68: 1},\n",
              " 'gp120': {1: 1, 64: 1},\n",
              " 'could': {43: 1, 67: 1},\n",
              " 'all': {24: 1, 15: 1, 67: 1},\n",
              " 'trigger': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'kinases': {33: 1, 67: 1},\n",
              " 'p56lck': {1: 1, 64: 1},\n",
              " 'p59fyn': {1: 1, 64: 1},\n",
              " 'phosphorylation': {1: 1, 67: 1},\n",
              " 'Shc': {1: 1, 63: 1, 65: 1},\n",
              " 'adaptor': {1: 1, 67: 1},\n",
              " 'Ras': {33: 1, 1: 1, 63: 1, 65: 1},\n",
              " 'they': {29: 1, 67: 1},\n",
              " 'differed': {19: 1, 35: 1, 67: 1},\n",
              " 'significantly': {30: 1, 67: 1},\n",
              " 'their': {22: 1, 67: 1},\n",
              " 'ability': {1: 1, 67: 1},\n",
              " 'Lack': {1: 1, 63: 1, 65: 1},\n",
              " 'full': {0: 1, 67: 1},\n",
              " 'correlated': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'dramatically': {30: 1, 67: 1},\n",
              " 'reduced': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'capacity': {1: 1, 67: 1},\n",
              " 'calcium': {1: 1, 67: 1},\n",
              " 'flux': {1: 1, 67: 1},\n",
              " 'complemented': {19: 1, 35: 1, 67: 1},\n",
              " 'ionophore': {1: 1, 67: 1},\n",
              " 'results': {33: 1, 12: 1, 67: 1},\n",
              " 'distinct': {0: 1, 67: 1},\n",
              " 'involved': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'Ras/protein': {1: 1, 49: 1, 63: 1, 65: 1},\n",
              " 'C': {1: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'pathways': {33: 1, 67: 1},\n",
              " 'Ligand-dependent': {0: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'repression': {1: 1, 67: 1},\n",
              " 'erythroid': {0: 1, 67: 1},\n",
              " 'GATA-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'estrogen': {1: 1, 67: 1},\n",
              " 'High-dose': {0: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'administration': {1: 1, 67: 1},\n",
              " 'anemia': {1: 1, 67: 1},\n",
              " 'mammals': {33: 1, 67: 1},\n",
              " 'chickens': {33: 1, 67: 1},\n",
              " 'estrogens': {33: 1, 67: 1},\n",
              " 'stimulate': {44: 1, 38: 1, 67: 1},\n",
              " 'outgrowth': {1: 1, 67: 1},\n",
              " 'bone': {1: 1, 67: 1},\n",
              " 'marrow-derived': {0: 1, 56: 1, 68: 1},\n",
              " 'progenitor': {1: 1, 67: 1},\n",
              " 'delay': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'maturation': {1: 1, 67: 1},\n",
              " 'down-regulation': {1: 1, 56: 1, 68: 1},\n",
              " 'many': {0: 1, 67: 1},\n",
              " 'cell-specific': {0: 1, 56: 1, 68: 1},\n",
              " 'genes': {33: 1, 67: 1},\n",
              " 'including': {4: 1, 67: 1},\n",
              " 'alpha-': {1: 1, 56: 1, 68: 1},\n",
              " 'beta-globin': {0: 1, 1: 1, 23: 1, 56: 1, 68: 1},\n",
              " 'band': {1: 1, 67: 1},\n",
              " '3': {42: 1, 46: 1, 48: 1},\n",
              " '4.1': {42: 1, 54: 1},\n",
              " 'histone': {1: 1, 67: 1},\n",
              " 'H5': {1: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'here': {30: 1, 67: 1},\n",
              " 'also': {30: 1, 67: 1},\n",
              " 'reduce': {44: 1, 38: 1, 67: 1},\n",
              " 'number': {1: 1, 67: 1},\n",
              " 'marrow': {0: 1, 1: 1, 67: 1},\n",
              " 'cultures': {33: 1, 67: 1},\n",
              " 'To': {5: 1, 63: 1, 65: 1},\n",
              " 'address': {44: 1, 38: 1, 67: 1},\n",
              " 'potential': {0: 1, 1: 1, 67: 1},\n",
              " 'mechanisms': {33: 1, 67: 1},\n",
              " 'suppress': {44: 1, 38: 1, 67: 1},\n",
              " 'erythropoiesis': {1: 1, 67: 1},\n",
              " 'examined': {19: 1, 35: 1, 67: 1},\n",
              " 'participates': {12: 1, 67: 1},\n",
              " 'regulation': {1: 1, 67: 1},\n",
              " 'majority': {1: 1, 67: 1},\n",
              " 'necessary': {0: 1, 67: 1},\n",
              " 'erythrocytes': {33: 1, 67: 1},\n",
              " 'transcriptional': {0: 1, 67: 1},\n",
              " 'strongly': {30: 1, 67: 1},\n",
              " 'repressed': {19: 1, 35: 1, 67: 1},\n",
              " 'ER': {1: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'ligand-dependent': {0: 1, 56: 1, 68: 1},\n",
              " 'manner': {1: 1, 67: 1},\n",
              " 'reversible': {0: 1, 67: 1},\n",
              " 'presence': {1: 1, 67: 1},\n",
              " '4-hydroxytamoxifen': {1: 1, 56: 1, 64: 1, 68: 1},\n",
              " 'ER-mediated': {0: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'occurs': {12: 1, 67: 1},\n",
              " 'artificial': {0: 1, 67: 1},\n",
              " 'promoter': {0: 1, 1: 1, 67: 1},\n",
              " 'containing': {1: 1, 4: 1, 67: 1},\n",
              " 'GATA-binding': {0: 1, 1: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'as': {23: 1, 30: 1, 67: 1},\n",
              " 'well': {1: 1, 30: 1, 67: 1},\n",
              " 'context': {1: 1, 67: 1},\n",
              " 'intact': {0: 1, 67: 1},\n",
              " 'normally': {30: 1, 67: 1},\n",
              " 'regulated': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'bind': {44: 1, 38: 1, 67: 1},\n",
              " 'each': {24: 1, 67: 1},\n",
              " 'other': {0: 1, 24: 1, 67: 1},\n",
              " 'vitro': {18: 1, 67: 1},\n",
              " 'absence': {1: 1, 67: 1},\n",
              " 'DNA': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'coimmunoprecipitation': {1: 1, 67: 1},\n",
              " 'experiments': {33: 1, 67: 1},\n",
              " 'COS': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'associate': {44: 1, 38: 1, 67: 1},\n",
              " 'Mapping': {1: 1, 63: 1, 65: 1},\n",
              " 'indicate': {44: 1, 38: 1, 67: 1},\n",
              " 'form': {1: 1, 23: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'at': {23: 1, 67: 1},\n",
              " 'least': {41: 1, 8: 1, 67: 1},\n",
              " 'contacts': {33: 1, 67: 1},\n",
              " 'finger': {1: 1, 67: 1},\n",
              " 'N-terminal': {0: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'domain': {0: 1, 1: 1, 67: 1},\n",
              " 'speculate': {44: 1, 38: 1, 67: 1},\n",
              " 'exert': {44: 1, 38: 1, 35: 1, 67: 1},\n",
              " 'modulating': {0: 1, 1: 1, 4: 1, 67: 1},\n",
              " 'protein-protein': {0: 1, 56: 1, 68: 1},\n",
              " 'interaction': {1: 1, 67: 1},\n",
              " 'ABSTRACT': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'TRUNCATED': {35: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'AT': {1: 1, 23: 1, 58: 1, 63: 1, 65: 1},\n",
              " '250': {42: 1, 48: 1},\n",
              " 'WORDS': {33: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'Mouse': {1: 1, 63: 1, 65: 1},\n",
              " 'alpha': {1: 1, 67: 1},\n",
              " 'Interleukin-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 68: 1},\n",
              " 'control': {0: 1, 1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'shown': {35: 1, 67: 1},\n",
              " 'interleukin-1': {1: 1, 56: 1, 64: 1, 68: 1},\n",
              " 'IL-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'IL-2R': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'CD4-CD8- murine': {0: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'lymphocyte': {0: 1, 1: 1, 67: 1},\n",
              " 'precursors': {33: 1, 67: 1},\n",
              " 'Here': {30: 1, 63: 1, 65: 1},\n",
              " 'map': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'interleukin': {1: 1, 67: 1},\n",
              " 'responsiveness': {1: 1, 67: 1},\n",
              " 'mouse': {1: 1, 67: 1},\n",
              " 'thymic': {0: 1, 67: 1},\n",
              " 'lymphoma-derived': {0: 1, 56: 1, 68: 1},\n",
              " 'hybridoma': {1: 1, 67: 1},\n",
              " 'PC60': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'response': {1: 1, 38: 1, 67: 1},\n",
              " '+': {0: 1, 36: 1, 69: 1},\n",
              " 'biphasic': {0: 1, 67: 1},\n",
              " 'synthesis-independent': {0: 1, 56: 1, 68: 1},\n",
              " 'appearance': {1: 1, 67: 1},\n",
              " 'mRNA': {1: 1, 63: 1, 66: 1},\n",
              " 'blocked': {19: 1, 35: 1, 67: 1},\n",
              " 'inhibitors': {33: 1, 67: 1},\n",
              " 'It': {29: 1, 63: 1, 65: 1},\n",
              " 'primes': {12: 1, 67: 1},\n",
              " 'become': {44: 1, 38: 1, 35: 1, 67: 1},\n",
              " 'responsive': {0: 1, 1: 1, 67: 1},\n",
              " 'thereby': {30: 1, 67: 1},\n",
              " 'prepares': {12: 1, 67: 1},\n",
              " 'second': {0: 1, 1: 1, 30: 1, 67: 1},\n",
              " 'phase': {1: 1, 67: 1},\n",
              " '100-fold': {30: 1, 56: 1, 64: 1, 68: 1},\n",
              " 'further': {0: 1, 25: 1, 30: 1, 67: 1},\n",
              " 'increase': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'transcripts': {33: 1, 67: 1},\n",
              " 'Transient': {0: 1, 63: 1, 65: 1},\n",
              " 'several': {0: 1, 67: 1},\n",
              " 'promoter-proximal': {0: 1, 56: 1, 68: 1},\n",
              " 'contribute': {44: 1, 38: 1, 67: 1},\n",
              " 'most': {41: 1, 8: 1, 0: 1, 67: 1},\n",
              " 'importantly': {30: 1, 67: 1},\n",
              " 'hand': {1: 1, 67: 1},\n",
              " 'depends': {12: 1, 67: 1},\n",
              " '78-nucleotide': {0: 1, 56: 1, 64: 1, 68: 1},\n",
              " 'segment': {1: 1, 67: 1},\n",
              " '1.3': {42: 1, 54: 1},\n",
              " 'kilobases': {33: 1, 67: 1},\n",
              " 'start': {1: 1, 38: 1, 67: 1},\n",
              " 'IL-2-inducible': {0: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'lies': {12: 1, 67: 1},\n",
              " 'becomes': {12: 1, 67: 1},\n",
              " 'DNase': {1: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'hypersensitive': {0: 1, 67: 1},\n",
              " 'normal': {0: 1, 67: 1},\n",
              " 'induced': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'three': {42: 1, 67: 1},\n",
              " 'Two': {42: 1, 63: 1, 65: 1},\n",
              " 'STAT': {33: 1, 1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'Hematopoietic': {0: 1, 63: 1, 65: 1},\n",
              " 'lineage': {1: 1, 67: 1},\n",
              " 'commitment': {1: 1, 67: 1},\n",
              " 'factors': {33: 1, 1: 1, 67: 1},\n",
              " 'review': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'focuses': {12: 1, 67: 1},\n",
              " 'roles': {33: 1, 67: 1},\n",
              " 'hematopoietic': {0: 1, 67: 1},\n",
              " 'A': {1: 1, 24: 1, 3: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'brief': {0: 1, 67: 1},\n",
              " 'introduction': {1: 1, 67: 1},\n",
              " 'asymmetric': {0: 1, 67: 1},\n",
              " 'division': {1: 1, 67: 1},\n",
              " 'discussion': {1: 1, 67: 1},\n",
              " 'methods': {33: 1, 67: 1},\n",
              " 'important': {0: 1, 67: 1},\n",
              " 'specifying': {4: 1, 67: 1},\n",
              " 'types': {33: 1, 67: 1},\n",
              " 'Next': {30: 1, 63: 1, 65: 1},\n",
              " 'presented': {19: 1, 35: 1, 67: 1},\n",
              " 'use': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'embryonic': {0: 1, 67: 1},\n",
              " 'stem': {1: 1, 38: 1, 67: 1},\n",
              " 'analysis': {1: 1, 67: 1},\n",
              " 'targeted': {19: 1, 35: 1, 67: 1},\n",
              " 'disruption': {1: 1, 67: 1},\n",
              " 'analyze': {44: 1, 38: 1, 67: 1},\n",
              " 'hematopoiesis': {1: 1, 67: 1},\n",
              " 'Finally': {30: 1, 63: 1, 65: 1},\n",
              " 'status': {1: 1, 67: 1},\n",
              " 'our': {22: 1, 67: 1},\n",
              " 'current': {0: 1, 1: 1, 67: 1},\n",
              " 'knowledge': {1: 1, 67: 1},\n",
              " 'concerning': {4: 1, 67: 1},\n",
              " 'myeloid': {0: 1, 1: 1, 67: 1},\n",
              " 'lymphoid': {0: 1, 67: 1},\n",
              " 'summarized': {35: 1, 67: 1},\n",
              " 'Epstein-Barr': {0: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'replicative': {0: 1, 1: 1, 67: 1},\n",
              " 'de': {18: 1, 67: 1},\n",
              " 'novo': {18: 1, 67: 1},\n",
              " 'thymocytes': {33: 1, 67: 1},\n",
              " 'simultaneous': {0: 1, 67: 1},\n",
              " 'early': {0: 1, 30: 1, 67: 1},\n",
              " 'BZLF-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'its': {22: 1, 67: 1},\n",
              " 'repressor': {1: 1, 67: 1},\n",
              " 'RAZ': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'EBV': {0: 1, 1: 1, 33: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'known': {0: 1, 35: 1, 67: 1},\n",
              " 'infect': {44: 1, 67: 1},\n",
              " 'epithelial': {0: 1, 1: 1, 67: 1},\n",
              " 'others': {33: 1, 67: 1},\n",
              " 'can': {43: 1, 67: 1},\n",
              " 'subset': {1: 1, 67: 1},\n",
              " 'Infection': {1: 1, 63: 1, 65: 1},\n",
              " 'accompanied': {19: 1, 35: 1, 67: 1},\n",
              " 'linear': {0: 1, 1: 1, 67: 1},\n",
              " 'genome': {1: 1, 67: 1},\n",
              " '8': {42: 1, 48: 1},\n",
              " 'hr': {1: 1, 67: 1},\n",
              " 'Circularization': {1: 1, 63: 1, 65: 1},\n",
              " 'detected': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'contrast': {1: 1, 67: 1},\n",
              " 'where': {26: 1, 67: 1},\n",
              " 'circularize': {44: 1, 67: 1},\n",
              " '24': {42: 1, 48: 1},\n",
              " 'BamHI': {1: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'ZLF-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'product': {1: 1, 67: 1},\n",
              " 'ZEBRA': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'RT-PCR': {1: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'observed': {19: 1, 35: 1, 67: 1},\n",
              " 'fusion': {1: 1, 67: 1},\n",
              " 'transcript': {1: 1, 67: 1},\n",
              " 'comprised': {19: 1, 35: 1, 67: 1},\n",
              " 'locus': {1: 1, 67: 1},\n",
              " 'BRLF-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'immunoprecipitation': {1: 1, 67: 1},\n",
              " 'addition': {1: 1, 67: 1},\n",
              " 'EBNA-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'transcribed': {19: 1, 35: 1, 67: 1},\n",
              " 'Fp': {1: 1, 63: 1, 65: 1},\n",
              " 'rather': {30: 1, 67: 1},\n",
              " 'than': {23: 1, 67: 1},\n",
              " 'Cp/Wp': {1: 1, 49: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'latently': {30: 1, 67: 1},\n",
              " 'Transcripts': {33: 1, 63: 1, 65: 1},\n",
              " 'encoding': {0: 1, 1: 1, 4: 1, 67: 1},\n",
              " 'gp350/220': {1: 1, 49: 1, 57: 1, 64: 1},\n",
              " 'coat': {1: 1, 67: 1},\n",
              " 'did': {19: 1, 67: 1},\n",
              " 'find': {44: 1, 38: 1, 67: 1},\n",
              " 'any': {24: 1, 67: 1},\n",
              " 'evidence': {1: 1, 67: 1},\n",
              " 'LMP-2A': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'EBER-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'loci': {33: 1, 67: 1},\n",
              " 'observations': {33: 1, 67: 1},\n",
              " 'differs': {12: 1, 67: 1},\n",
              " 'main': {0: 1, 67: 1},\n",
              " 'difference': {1: 1, 67: 1},\n",
              " 'ever': {30: 1, 67: 1},\n",
              " 'circularizes': {12: 1, 67: 1},\n",
              " 'Rather': {30: 1, 63: 1, 65: 1},\n",
              " 'remains': {12: 1, 67: 1},\n",
              " 'configuration': {1: 1, 67: 1},\n",
              " 'Identification': {1: 1, 63: 1, 65: 1},\n",
              " 'purification': {1: 1, 67: 1},\n",
              " 'Stat': {33: 1, 1: 1, 63: 1, 65: 1},\n",
              " 'activated': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'key': {0: 1, 67: 1},\n",
              " 'cytokine': {1: 1, 67: 1},\n",
              " 'immune': {0: 1, 67: 1},\n",
              " 'Following': {4: 1, 63: 1, 65: 1},\n",
              " 'various': {0: 1, 67: 1},\n",
              " 'chains': {33: 1, 67: 1},\n",
              " 'transcriptionally': {30: 1, 67: 1},\n",
              " 'turn': {1: 1, 67: 1},\n",
              " 'secreted': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'serves': {12: 1, 67: 1},\n",
              " 'differentiation': {1: 1, 67: 1},\n",
              " 'Several': {0: 1, 63: 1, 65: 1},\n",
              " 'recent': {0: 1, 67: 1},\n",
              " 'studies': {33: 1, 12: 1, 67: 1},\n",
              " 'implicated': {19: 1, 35: 1, 67: 1},\n",
              " 'Jak': {1: 1, 63: 1, 65: 1},\n",
              " 'lead': {19: 1, 1: 1, 35: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'set': {19: 1, 1: 1, 38: 1, 35: 1, 67: 1},\n",
              " 'out': {23: 1, 30: 1, 40: 1, 67: 1},\n",
              " 'contain': {44: 1, 38: 1, 67: 1},\n",
              " 'activities': {33: 1, 67: 1},\n",
              " 'Similar': {0: 1, 63: 1, 65: 1},\n",
              " 'transformed': {0: 1, 19: 1, 35: 1, 67: 1},\n",
              " 'line': {1: 1, 67: 1},\n",
              " 'termed': {19: 1, 35: 1, 67: 1},\n",
              " 'YT': {1: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'purified': {0: 1, 19: 1, 35: 1, 67: 1},\n",
              " 'principal': {0: 1, 67: 1},\n",
              " 'component': {1: 1, 67: 1},\n",
              " 'bears': {12: 1, 67: 1},\n",
              " 'relatedness': {1: 1, 67: 1},\n",
              " 'prolactin-induced': {0: 1, 56: 1, 68: 1},\n",
              " 'sheep': {1: 1, 67: 1},\n",
              " 'mammary': {0: 1, 67: 1},\n",
              " 'gland': {1: 1, 67: 1},\n",
              " 'tissue': {1: 1, 67: 1},\n",
              " 'hypothesize': {38: 1, 67: 1},\n",
              " 'hStat5': {1: 1, 63: 1, 64: 1, 66: 1},\n",
              " 'helps': {12: 1, 67: 1},\n",
              " 'govern': {44: 1, 38: 1, 67: 1},\n",
              " 'biological': {0: 1, 67: 1},\n",
              " 'E2F-1': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'cyclin-like': {0: 1, 56: 1, 68: 1},\n",
              " 'repair': {1: 1, 67: 1},\n",
              " 'enzyme': {1: 1, 67: 1},\n",
              " 'uracil-DNA': {1: 1, 56: 1, 63: 1, 66: 1, 68: 1},\n",
              " 'glycosylase': {1: 1, 67: 1},\n",
              " 'autoregulatory': {0: 1, 1: 1, 67: 1},\n",
              " 'mechanism': {1: 1, 67: 1},\n",
              " 'cycle-dependent': {0: 1, 56: 1, 68: 1},\n",
              " 'regulates': {12: 1, 67: 1},\n",
              " 'species': {33: 1, 67: 1},\n",
              " 'UDG': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'osteosarcoma': {1: 1, 67: 1},\n",
              " 'Saos-2': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 68: 1},\n",
              " 'deletion': {1: 1, 67: 1},\n",
              " 'sequences': {33: 1, 12: 1, 67: 1},\n",
              " 'activates': {12: 1, 67: 1},\n",
              " 'E2F': {1: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'putative': {0: 1, 67: 1},\n",
              " 'downstream': {0: 1, 30: 1, 67: 1},\n",
              " 'located': {0: 1, 67: 1},\n",
              " 'exon': {1: 1, 67: 1},\n",
              " 'E2F-1/DP1': {1: 1, 49: 1, 56: 1, 57: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'functional': {0: 1, 67: 1},\n",
              " 'relationship': {1: 1, 67: 1},\n",
              " 'between': {23: 1, 67: 1},\n",
              " 'High': {0: 1, 63: 1, 65: 1},\n",
              " 'assay': {1: 1, 44: 1, 67: 1},\n",
              " 'result': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'E2F-mediated': {0: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'Overexpression': {1: 1, 63: 1, 65: 1},\n",
              " 'Saos': {1: 1, 63: 1, 65: 1},\n",
              " 'growth': {1: 1, 67: 1},\n",
              " 'late': {0: 1, 30: 1, 67: 1},\n",
              " 'G1': {1: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'transiently': {30: 1, 67: 1},\n",
              " 'arrest': {1: 1, 44: 1, 67: 1},\n",
              " 'progressing': {4: 1, 67: 1},\n",
              " 'S': {1: 1, 3: 1, 58: 1, 63: 1, 65: 1},\n",
              " 'hypothetical': {0: 1, 67: 1},\n",
              " 'integrates': {12: 1, 67: 1},\n",
              " 'one': {1: 1, 42: 1, 29: 1, 67: 1},\n",
              " 'cycle': {1: 1, 67: 1},\n",
              " 'likely': {0: 1, 30: 1, 67: 1},\n",
              " 'implicates': {12: 1, 67: 1},\n",
              " 'multifunctional': {0: 1, 67: 1},\n",
              " 'enzymes': {33: 1, 67: 1},\n",
              " 'possibly': {30: 1, 67: 1},\n",
              " 'damage': {1: 1, 67: 1},\n",
              " 'negative': {0: 1, 67: 1},\n",
              " 'Cellular': {0: 1, 63: 1, 65: 1},\n",
              " 'molecular': {0: 1, 1: 1, 67: 1},\n",
              " 'IFN-gamma': {2: 1, 1: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'IL-12': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'Interferon-gamma': {1: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'immunoregulatory': {0: 1, 67: 1},\n",
              " 'produced': {19: 1, 35: 1, 67: 1},\n",
              " 'predominantly': {30: 1, 67: 1},\n",
              " 'large': {0: 1, 67: 1},\n",
              " 'granular': {0: 1, 67: 1},\n",
              " 'LGL': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'extracellular': {0: 1, 67: 1},\n",
              " 'particular': {0: 1, 67: 1},\n",
              " 'interleukins': {33: 1, 67: 1},\n",
              " 'ILs': {33: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'potent': {0: 1, 67: 1},\n",
              " 'inducers': {33: 1, 67: 1},\n",
              " 'Although': {23: 1, 63: 1, 65: 1},\n",
              " 'it': {29: 1, 67: 1},\n",
              " 'reported': {19: 1, 35: 1, 67: 1},\n",
              " 'there': {34: 1, 30: 1, 67: 1},\n",
              " 'some': {24: 1, 67: 1},\n",
              " 'produce': {44: 1, 38: 1, 67: 1},\n",
              " 'yet': {30: 1, 36: 1, 67: 1},\n",
              " 'report': {1: 1, 38: 1, 67: 1},\n",
              " 'killer': {1: 1, 67: 1},\n",
              " 'responds': {12: 1, 67: 1},\n",
              " 'similar': {0: 1, 67: 1},\n",
              " 'present': {0: 1, 1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'NK3.3': {1: 1, 54: 1, 59: 1, 60: 1, 62: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'derived': {0: 1, 19: 1, 35: 1, 67: 1},\n",
              " 'measured': {19: 1, 35: 1, 67: 1},\n",
              " 'increases': {33: 1, 12: 1, 67: 1},\n",
              " 'granulocyte-macrophage': {0: 1, 1: 1, 56: 1, 68: 1},\n",
              " 'colony-stimulating': {0: 1, 56: 1, 68: 1},\n",
              " 'GM-CSF': {1: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'cytoplasmic': {0: 1, 67: 1},\n",
              " 'when': {26: 1, 67: 1},\n",
              " 'together': {30: 1, 67: 1},\n",
              " 'synergized': {19: 1, 67: 1},\n",
              " 'induction': {1: 1, 67: 1},\n",
              " 'synergy': {1: 1, 67: 1},\n",
              " 'attributed': {35: 1, 67: 1},\n",
              " 'increased': {19: 1, 35: 1, 67: 1},\n",
              " 'accumulation': {1: 1, 67: 1},\n",
              " 'stability': {1: 1, 67: 1},\n",
              " 'mRNAs': {33: 1, 63: 1, 66: 1},\n",
              " 'investigate': {44: 1, 38: 1, 67: 1},\n",
              " 'five': {42: 1, 67: 1},\n",
              " 'cyclosporin': {1: 1, 67: 1},\n",
              " 'CsA': {1: 1, 63: 1, 65: 1},\n",
              " 'transforming': {0: 1, 1: 1, 4: 1, 67: 1},\n",
              " 'factor-beta': {1: 1, 56: 1, 68: 1},\n",
              " 'cycloheximide': {1: 1, 67: 1},\n",
              " 'genistein': {1: 1, 67: 1},\n",
              " 'staurosporine': {1: 1, 67: 1},\n",
              " 'analyzing': {4: 1, 67: 1},\n",
              " 'new': {0: 1, 67: 1},\n",
              " 'synthesis': {1: 1, 67: 1},\n",
              " 'appears': {12: 1, 67: 1},\n",
              " 'only': {0: 1, 30: 1, 67: 1},\n",
              " 'partially': {30: 1, 67: 1},\n",
              " 'depend': {44: 1, 38: 1, 67: 1},\n",
              " 'Furthermore': {30: 1, 63: 1, 65: 1},\n",
              " 'inhibitor': {1: 1, 67: 1},\n",
              " 'generally': {30: 1, 67: 1},\n",
              " 'inactive': {0: 1, 67: 1},\n",
              " 'suppression': {1: 1, 67: 1},\n",
              " 'agents': {33: 1, 67: 1},\n",
              " 'independent': {0: 1, 67: 1},\n",
              " 'inhibition': {1: 1, 67: 1},\n",
              " 'AP1': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'complexes': {33: 1, 67: 1},\n",
              " 'leading': {4: 1, 67: 1},\n",
              " 'serve': {44: 1, 38: 1, 67: 1},\n",
              " 'dissecting': {4: 1, 67: 1},\n",
              " 'biochemical': {0: 1, 67: 1},\n",
              " 'events': {33: 1, 67: 1},\n",
              " 'p95vav': {1: 1, 64: 1},\n",
              " 'Stimulation': {1: 1, 63: 1, 65: 1},\n",
              " 'TCR': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'multiple': {0: 1, 67: 1},\n",
              " 'numerous': {0: 1, 67: 1},\n",
              " 'substrates': {33: 1, 67: 1},\n",
              " 'One': {42: 1, 63: 1, 65: 1},\n",
              " 'substrate': {0: 1, 1: 1, 67: 1},\n",
              " 'expressed': {19: 1, 35: 1, 67: 1},\n",
              " 'exclusively': {30: 1, 67: 1},\n",
              " 'trophoblast': {1: 1, 67: 1},\n",
              " 'contains': {12: 1, 67: 1},\n",
              " 'structural': {0: 1, 67: 1},\n",
              " 'motifs': {33: 1, 67: 1},\n",
              " 'Src': {1: 1, 63: 1, 65: 1},\n",
              " 'homology': {1: 1, 67: 1},\n",
              " 'pleckstrin': {1: 1, 67: 1},\n",
              " 'domains': {33: 1, 67: 1},\n",
              " 'guanine': {0: 1, 1: 1, 67: 1},\n",
              " 'nucleotide': {0: 1, 1: 1, 67: 1},\n",
              " 'exchange': {1: 1, 67: 1},\n",
              " 'TCR-mediated': {0: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'processes': {33: 1, 12: 1, 67: 1},\n",
              " 'unclear': {0: 1, 67: 1},\n",
              " 'overexpression': {1: 1, 67: 1},\n",
              " 'alone': {30: 1, 67: 1},\n",
              " 'Jurkat': {1: 1, 63: 1, 65: 1},\n",
              " 'NFAT': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'synergizes': {12: 1, 67: 1},\n",
              " 'inducing': {1: 1, 4: 1, 67: 1},\n",
              " 'NFAT-': {1: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'interleukin-2-dependent': {0: 1, 56: 1, 64: 1, 68: 1},\n",
              " 'G-protein-coupled': {0: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'modulated': {0: 1, 19: 1, 35: 1, 67: 1},\n",
              " 'suggesting': {4: 1, 67: 1},\n",
              " 'removal': {1: 1, 67: 1},\n",
              " '67': {42: 1, 48: 1},\n",
              " 'amino': {1: 1, 67: 1},\n",
              " 'acids': {33: 1, 67: 1},\n",
              " 'NIH': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " '3T3': {1: 1, 63: 1, 64: 1, 66: 1},\n",
              " 'p95vav-induced': {0: 1, 56: 1, 64: 1, 68: 1},\n",
              " 'mimicked': {19: 1, 35: 1, 67: 1},\n",
              " 'though': {23: 1, 30: 1, 67: 1},\n",
              " 'upon': {23: 1, 67: 1},\n",
              " 'Raf': {1: 1, 63: 1, 65: 1},\n",
              " 'FK506': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'calcineurin': {1: 1, 67: 1},\n",
              " 'dissect': {44: 1, 67: 1},\n",
              " 'involvement': {1: 1, 67: 1},\n",
              " 'analyzed': {19: 1, 35: 1, 67: 1},\n",
              " 'mutants': {33: 1, 67: 1},\n",
              " 'deficient': {0: 1, 1: 1, 67: 1},\n",
              " 'showed': {19: 1, 35: 1, 67: 1},\n",
              " 'However': {30: 1, 26: 1, 63: 1, 65: 1},\n",
              " 'does': {12: 1, 67: 1},\n",
              " 'appear': {44: 1, 38: 1, 67: 1},\n",
              " 'influence': {1: 1, 44: 1, 38: 1, 67: 1},\n",
              " 'TCR-induced': {0: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'free': {0: 1, 67: 1},\n",
              " 'Taken': {35: 1, 63: 1, 65: 1},\n",
              " 'plays': {12: 1, 67: 1},\n",
              " 'unidentified': {0: 1, 67: 1},\n",
              " 'proximal': {0: 1, 1: 1, 67: 1},\n",
              " 'position': {1: 1, 67: 1},\n",
              " 'Positive': {0: 1, 63: 1, 65: 1},\n",
              " 'AML1-related': {0: 1, 56: 1, 63: 1, 64: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'PEBP2': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'consensus': {1: 1, 67: 1},\n",
              " 'polyomavirus': {1: 1, 67: 1},\n",
              " 'binding-protein': {1: 1, 56: 1, 68: 1},\n",
              " 'consists': {12: 1, 67: 1},\n",
              " 'beta': {1: 1, 67: 1},\n",
              " 'subunits': {33: 1, 67: 1},\n",
              " 'There': {34: 1, 63: 1, 65: 1},\n",
              " 'subunit': {1: 1, 67: 1},\n",
              " 'homologue': {1: 1, 67: 1},\n",
              " 'AML1': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'breakpoints': {33: 1, 67: 1},\n",
              " 't': {1: 1, 67: 1},\n",
              " ';': {45: 1, 69: 1},\n",
              " '21': {42: 1, 48: 1},\n",
              " 'leukemias': {33: 1, 1: 1, 67: 1},\n",
              " 'A1': {1: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'A-gene': {0: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'B1': {1: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'B2': {1: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'B-encoded': {0: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'isomers': {33: 1, 67: 1},\n",
              " 'bound': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'promoter-driven': {0: 1, 56: 1, 68: 1},\n",
              " 'reporter': {1: 1, 67: 1},\n",
              " 'plasmid': {1: 1, 67: 1},\n",
              " 'unstimulated': {0: 1, 67: 1},\n",
              " '12-O-tetradecanoylphorbol': {1: 1, 56: 1, 63: 1, 64: 1, 66: 1, 68: 1},\n",
              " '13-acetate/phytohemagglutinin-stimulated': {0: 1,\n",
              "  49: 1,\n",
              "  56: 1,\n",
              "  57: 1,\n",
              "  64: 1,\n",
              "  68: 1},\n",
              " 'suppressed': {19: 1, 35: 1, 67: 1},\n",
              " 'Coexpression': {1: 1, 63: 1, 65: 1},\n",
              " 'determined': {19: 1, 35: 1, 67: 1},\n",
              " 'B1/alpha': {1: 1, 49: 1, 57: 1, 63: 1, 64: 1, 65: 1},\n",
              " 'ratio': {1: 1, 67: 1},\n",
              " 'extract': {1: 1, 67: 1},\n",
              " 'contained': {19: 1, 35: 1, 67: 1},\n",
              " 'site-binding': {0: 1, 56: 1, 68: 1},\n",
              " 'cross-reacted': {19: 1, 56: 1, 68: 1},\n",
              " 'antimouse': {0: 1, 67: 1},\n",
              " 'Northern': {0: 1, 1: 1, 63: 1, 65: 1},\n",
              " 'blot': {1: 1, 67: 1},\n",
              " 'indicated': {19: 1, 35: 1, 67: 1},\n",
              " 'determine': {44: 1, 38: 1, 35: 1, 67: 1},\n",
              " 'precise': {0: 1, 67: 1},\n",
              " 'suggested': {19: 1, 35: 1, 67: 1},\n",
              " 'importance': {1: 1, 67: 1},\n",
              " 'relative': {0: 1, 1: 1, 67: 1},\n",
              " 'isoforms': {33: 1, 12: 1, 67: 1},\n",
              " 'regulating': {1: 1, 4: 1, 67: 1},\n",
              " 'priming': {0: 1, 1: 1, 4: 1, 67: 1},\n",
              " 'enhances': {12: 1, 67: 1},\n",
              " 'LPS-induced': {0: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'TNF': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'augmenting': {4: 1, 67: 1},\n",
              " 'MRNA': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " '/macrophages': {33: 1, 49: 1},\n",
              " 'bacterial': {0: 1, 67: 1},\n",
              " 'endotoxin': {1: 1, 67: 1},\n",
              " 'lipopolysaccharide': {1: 1, 67: 1},\n",
              " 'critical': {0: 1, 67: 1},\n",
              " 'highly': {30: 1, 67: 1},\n",
              " 'defence': {1: 1, 67: 1},\n",
              " 'augmentation': {1: 1, 67: 1},\n",
              " 'LPS': {33: 1, 1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'responses': {33: 1, 67: 1},\n",
              " 'interferon': {1: 1, 67: 1},\n",
              " 'gamma': {1: 1, 44: 1, 67: 1},\n",
              " 'referred': {35: 1, 67: 1},\n",
              " 'established': {0: 1, 19: 1, 35: 1, 67: 1},\n",
              " 'poorly': {30: 1, 67: 1},\n",
              " 'Using': {4: 1, 63: 1, 65: 1},\n",
              " 'tumour': {0: 1, 1: 1, 67: 1},\n",
              " 'necrosis': {1: 1, 67: 1},\n",
              " 'designed': {35: 1, 67: 1},\n",
              " 'analyse': {44: 1, 38: 1, 67: 1},\n",
              " 'detail': {1: 1, 67: 1},\n",
              " 'Priming': {1: 1, 4: 1, 63: 1, 65: 1},\n",
              " 'manifested': {35: 1, 67: 1},\n",
              " 'level': {1: 1, 67: 1},\n",
              " 'pre-treatment': {1: 1, 56: 1, 68: 1},\n",
              " 'affected': {19: 1, 0: 1, 35: 1, 67: 1},\n",
              " 'magnitude': {1: 1, 67: 1},\n",
              " 'sensitivity': {1: 1, 67: 1},\n",
              " 'occurred': {19: 1, 35: 1, 67: 1},\n",
              " 'after': {23: 1, 30: 1, 67: 1},\n",
              " 'hours': {33: 1, 67: 1},\n",
              " 'treatment': {1: 1, 67: 1},\n",
              " 'primed': {0: 1, 19: 1, 35: 1, 67: 1},\n",
              " 'state': {1: 1, 67: 1},\n",
              " 'M-CSF': {1: 1, 56: 1, 63: 1, 65: 1, 66: 1, 68: 1},\n",
              " 'Primed': {35: 1, 63: 1, 65: 1},\n",
              " 'higher': {27: 1, 67: 1},\n",
              " 'rate': {1: 1, 67: 1},\n",
              " 'freshly': {30: 1, 67: 1},\n",
              " 'isolated': {19: 1, 35: 1, 67: 1},\n",
              " 'marked': {0: 1, 35: 1, 67: 1},\n",
              " 'factor-kappa': {1: 1, 56: 1, 68: 1},\n",
              " 'oligonucleotide': {0: 1, 1: 1, 67: 1},\n",
              " 'An': {24: 1, 63: 1, 65: 1},\n",
              " 'finding': {1: 1, 4: 1, 67: 1},\n",
              " 'much': {0: 1, 30: 1, 67: 1},\n",
              " 'more': {27: 1, 25: 1, 67: 1},\n",
              " 'unprimed': {0: 1, 67: 1},\n",
              " 'T1/2': {1: 1, 49: 1, 57: 1, 63: 1, 64: 1, 65: 1},\n",
              " '6-8-fold': {30: 1, 56: 1, 64: 1, 68: 1},\n",
              " 'duration': {1: 1, 67: 1},\n",
              " 'longer': {27: 1, 25: 1, 30: 1, 67: 1},\n",
              " 'being': {4: 1, 67: 1},\n",
              " 'greater': {27: 1, 67: 1},\n",
              " 'possessed': {19: 1, 35: 1, 67: 1},\n",
              " 'H-89': {1: 1, 56: 1, 63: 1, 64: 1, 65: 1, 68: 1},\n",
              " 'substantially': {30: 1, 67: 1},\n",
              " 'had': {19: 1, 67: 1},\n",
              " 'Myc-associated': {0: 1, 56: 1, 63: 1, 65: 1, 68: 1},\n",
              " 'zinc': {1: 1, 67: 1},\n",
              " 'developmental': {0: 1, 67: 1},\n",
              " 'minimal': {0: 1, 67: 1},\n",
              " 'mutagenesis': {1: 1, 67: 1},\n",
              " 'analyses': {33: 1, 67: 1},\n",
              " 'initiator-like': {0: 1, 56: 1, 68: 1},\n",
              " 'cap': {1: 1, 67: 1},\n",
              " 'Ets': {33: 1, 1: 1, 63: 1, 65: 1},\n",
              " 'MAZ': {1: 1, 58: 1, 63: 1, 65: 1, 66: 1},\n",
              " 'predominant': {0: 1, 67: 1},\n",
              " 'last': {0: 1, 67: 1},\n",
              " 'resembles': {12: 1, 67: 1},\n",
              " 'ME1a1': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'G3AG4AG3': {1: 1, 63: 1, 64: 1, 65: 1, 66: 1},\n",
              " 'motif': {1: 1, 67: 1},\n",
              " ...}"
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ],
      "source": [
        "word_features"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "l--AFQBEpD-s"
      },
      "source": [
        "### Creating Final Label and Feature Sets\n",
        "\n",
        "The SVM model requires the features to be given as a list of dicts and the labels as a list of values. The constructed dictionaries for the labels and features are processed to create such lists such that the i-th entry in the labels list and features list correspond to the same word (this was the entire purpose of using dictionaries in the previous sections)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "ruqT55q2piqQ"
      },
      "outputs": [],
      "source": [
        "y = []\n",
        "x = []\n",
        "for word in words_pos_dict.keys():\n",
        "  y.append(word_labels[word])\n",
        "  x.append(word_features[word])\n",
        "y = np.asarray(y)\n",
        "x = np.asarray(x)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DjxPX3Yk4PUs"
      },
      "source": [
        "### Tests"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {
        "id": "6GeGKdfh1xL-"
      },
      "outputs": [],
      "source": [
        "# make sure features constructed correctly\n",
        "unique_feat_entries = set().union(*[d.values() for d in word_features.values()])\n",
        "assert unique_feat_entries == {1} or len(unique_feat_entries) == 0, \"features contain entries other than 1. Note that if they are 0, they should not be included\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0q4bnfBMfYTU"
      },
      "source": [
        "## Training SVM Model\n",
        "\n",
        "We train an SVM model (using libsvm) to identify words that are related to the biological domain."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install libsvm-official"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-a8H-R0aJlX2",
        "outputId": "bfeea6ed-906f-42cd-8f52-840c6e5ba30f"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting libsvm-official\n",
            "  Downloading libsvm-official-3.30.0.tar.gz (38 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.8/dist-packages (from libsvm-official) (1.7.3)\n",
            "Requirement already satisfied: numpy<1.23.0,>=1.16.5 in /usr/local/lib/python3.8/dist-packages (from scipy->libsvm-official) (1.21.6)\n",
            "Building wheels for collected packages: libsvm-official\n",
            "  Building wheel for libsvm-official (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for libsvm-official: filename=libsvm_official-3.30.0-cp38-cp38-linux_x86_64.whl size=101392 sha256=6a81619ea273ba2ef6566649e706a70760c400547d5637c5f80520ad394b09eb\n",
            "  Stored in directory: /root/.cache/pip/wheels/cb/72/13/dde14ffd172a2ef7d7b8faf088be188ca85d105a169c5dd4f9\n",
            "Successfully built libsvm-official\n",
            "Installing collected packages: libsvm-official\n",
            "Successfully installed libsvm-official-3.30.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {
        "id": "qVhsxdXmRa9u"
      },
      "outputs": [],
      "source": [
        "from libsvm.svmutil import *\n",
        "from sklearn.model_selection import StratifiedKFold\n",
        "from sklearn.metrics import precision_score, recall_score"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5NAS-h2YXgue"
      },
      "source": [
        "### Dataset Split\n",
        "\n",
        "The data is split into a testing a training dataset. The training set is used to determine the best model parameters and train the model. The testing set is used to determine the performance of the trained model on unseen data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "id": "navARSgqYOcb"
      },
      "outputs": [],
      "source": [
        "from sklearn.model_selection import train_test_split"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "id": "MUR41TETYrvU"
      },
      "outputs": [],
      "source": [
        "x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "m-SdXzr_X-hD"
      },
      "source": [
        "### K-Fold Cross-validation\n",
        "\n",
        "We use 10-fold cross validation to determine the best model from the training data. Once the best model is determined, it is then evaluated on the testing data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DI36DC5USamu",
        "outputId": "962dd871-87e1-4a64-8937-522591f34f41"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Fold 0:\n",
            "Accuracy = 84.9462% (1343/1581) (classification)\n",
            "Precision = 0.7100840336134454\n",
            "Recall = 0.5\n",
            "\n",
            "Fold 1:\n",
            "Accuracy = 84.3137% (1333/1581) (classification)\n",
            "Precision = 0.68\n",
            "Recall = 0.5029585798816568\n",
            "\n",
            "Fold 2:\n",
            "Accuracy = 84.2505% (1332/1581) (classification)\n",
            "Precision = 0.6731517509727627\n",
            "Recall = 0.5118343195266272\n",
            "\n",
            "Fold 3:\n",
            "Accuracy = 82.4794% (1304/1581) (classification)\n",
            "Precision = 0.6196078431372549\n",
            "Recall = 0.46745562130177515\n",
            "\n",
            "Fold 4:\n",
            "Accuracy = 83.8077% (1325/1581) (classification)\n",
            "Precision = 0.6736401673640168\n",
            "Recall = 0.4749262536873156\n",
            "\n",
            "Fold 5:\n",
            "Accuracy = 86.401% (1366/1581) (classification)\n",
            "Precision = 0.7421875\n",
            "Recall = 0.56047197640118\n",
            "\n",
            "Fold 6:\n",
            "Accuracy = 85.5696% (1352/1580) (classification)\n",
            "Precision = 0.7165354330708661\n",
            "Recall = 0.5384615384615384\n",
            "\n",
            "Fold 7:\n",
            "Accuracy = 83.5443% (1320/1580) (classification)\n",
            "Precision = 0.6547619047619048\n",
            "Recall = 0.4881656804733728\n",
            "\n",
            "Fold 8:\n",
            "Accuracy = 83.5443% (1320/1580) (classification)\n",
            "Precision = 0.6598360655737705\n",
            "Recall = 0.47633136094674555\n",
            "\n",
            "Fold 9:\n",
            "Accuracy = 85.8861% (1357/1580) (classification)\n",
            "Precision = 0.722007722007722\n",
            "Recall = 0.5532544378698225\n",
            "\n",
            "Average Accuracy of model Across all Folds: 84.47429122731168\n",
            "Average Precision of model on Biological Entities Across all Folds: 0.6851812420501743\n",
            "Average Recalls of model on Biological Entities Across all Folds: 0.5073859768550035\n"
          ]
        }
      ],
      "source": [
        "skf = StratifiedKFold(n_splits=10, shuffle=True)\n",
        "skf.get_n_splits(x_train, y_train)\n",
        "\n",
        "acc = []\n",
        "precisions = []\n",
        "recalls = []\n",
        "param = svm_parameter('-s 0 -t 0 -c 2')\n",
        "for i, (train_idx, val_idx) in enumerate(skf.split(x_train, y_train)):\n",
        "  print(f'Fold {i}:')\n",
        "  prob = svm_problem(y_train[train_idx], x_train[train_idx])\n",
        "  m = svm_train(prob, param)\n",
        "  pred_label, pred_acc, pred_val = svm_predict(y_train[val_idx], x_train[val_idx], m)\n",
        "  prec = precision_score(y_train[val_idx], pred_label, average='binary')\n",
        "  rec = recall_score(y_train[val_idx], pred_label, average='binary')\n",
        "  print(f'Precision = {prec}')\n",
        "  print(f'Recall = {rec}')\n",
        "\n",
        "  acc.append(pred_acc[0])\n",
        "  precisions.append(prec)\n",
        "  recalls.append(rec)\n",
        "  print('')\n",
        "\n",
        "avg_acc = sum(acc) / len(acc)\n",
        "avg_prec = sum(precisions) / len(precisions)\n",
        "avg_rec = sum(recalls) / len(recalls)\n",
        "print(f'Average Accuracy of model Across all Folds: {avg_acc}')\n",
        "print(f'Average Precision of model on Biological Entities Across all Folds: {avg_prec}')\n",
        "print(f'Average Recalls of model on Biological Entities Across all Folds: {avg_rec}')"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Final Model and Testing\n",
        "\n",
        "We train the model with the best performance on all of the training data and then evaluate it on the test data."
      ],
      "metadata": {
        "id": "P1uDD8AWJRmJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "param = svm_parameter('-s 0 -t 0 -c 2')\n",
        "prob = svm_problem(y_train, x_train)\n",
        "m = svm_train(prob, param)"
      ],
      "metadata": {
        "id": "DlXq3cvqJQvR"
      },
      "execution_count": 23,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ewFE68leSppQ",
        "outputId": "50151010-5c86-49c0-cf9b-de5ed627260d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Accuracy = 84.4848% (5723/6774) (classification)\n",
            "Precision for biological entities = 0.6903648269410664\n",
            "Recall for biological entities = 0.5061728395061729\n",
            "Overall (micro) Precision = 0.8448479480366106\n",
            "Overall (micro) Recall = 0.8448479480366106\n"
          ]
        }
      ],
      "source": [
        "p_label, p_acc, p_val = svm_predict(y_test, x_test, m)\n",
        "prec_bio = precision_score(y_test, p_label, average='binary')\n",
        "rec_bio = recall_score(y_test, p_label, average='binary')\n",
        "prec_micro = precision_score(y_test, p_label, average='micro')\n",
        "rec_micro = recall_score(y_test, p_label, average='micro')\n",
        "\n",
        "print(f'Precision for biological entities = {prec_bio}')\n",
        "print(f'Recall for biological entities = {rec_bio}')\n",
        "\n",
        "print(f'Overall (micro) Precision = {prec_micro}')\n",
        "print(f'Overall (micro) Recall = {prec_micro}')"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "KyRrW8c-J27-"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "env",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0]"
    },
    "orig_nbformat": 4,
    "vscode": {
      "interpreter": {
        "hash": "72f86c46800bdf116255aecb4724685bd63dbc1069855c869a8e3dd17ec92b2e"
      }
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "416e4d9a63ab48818f451146351e3241": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_1c82c6e6139a4717a16323eafde603de",
              "IPY_MODEL_6ee5b482980c48948be2666771a47d83",
              "IPY_MODEL_e5659618745e42469a1722dbc639f058"
            ],
            "layout": "IPY_MODEL_cb317891831f453a97168a4bc71b51a5"
          }
        },
        "1c82c6e6139a4717a16323eafde603de": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_da8006e95e1d4c3ab4cb6062ebb2dd1e",
            "placeholder": "​",
            "style": "IPY_MODEL_ae5d5fcaa6944a528c08504872a666b2",
            "value": "100%"
          }
        },
        "6ee5b482980c48948be2666771a47d83": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_87cc8d887f504979b98bf467585f4364",
            "max": 2000,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_2684f161f67d463cb1dd9a539d14d6d9",
            "value": 2000
          }
        },
        "e5659618745e42469a1722dbc639f058": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_90d8ae28b880486cbff0f062dfc35174",
            "placeholder": "​",
            "style": "IPY_MODEL_9e93286e84b34eac9900e7ceca8923be",
            "value": " 2000/2000 [00:04&lt;00:00, 454.47it/s]"
          }
        },
        "cb317891831f453a97168a4bc71b51a5": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "da8006e95e1d4c3ab4cb6062ebb2dd1e": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "ae5d5fcaa6944a528c08504872a666b2": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "87cc8d887f504979b98bf467585f4364": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "2684f161f67d463cb1dd9a539d14d6d9": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "90d8ae28b880486cbff0f062dfc35174": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9e93286e84b34eac9900e7ceca8923be": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        }
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}